## Germline findings (CPSR)

* Based on a germline report of the query case using the [Cancer Predisposition Sequencing Reporter (CPSR)](https://github.com/sigven/cpsr), we here list the main findings from the report with respect
to variants in cancer predisposition genes
* Pathogenic, likely pathogenic variants, and variants of uncertain signinificance (VUS) are listed,
for variants with existing classifications in ClinVar, and for _novel_ variants (not found in ClinVar)

```{r prepare_cpsr_data, echo = F, eval = T}

cpsr_rep <- pcg_report[['content']][['cpsr']][['report']]

class543_germline_variants <- dplyr::bind_rows(
  cpsr_rep$content$snv_indel$disp$class5, 
  cpsr_rep$content$snv_indel$disp$class4, 
  cpsr_rep$content$snv_indel$disp$class3)

cpsr_clinvar_variants <- data.frame()
cpsr_other_variants <- data.frame()

if(NROW(class543_germline_variants) > 0){
  
  class543_germline_variants <- class543_germline_variants |>
    dplyr::mutate(GENOTYPE = as.character(GENOTYPE))
  
  cpsr_clinvar_variants <- class543_germline_variants |>
    dplyr::filter(CPSR_CLASSIFICATION_SOURCE == "ClinVar")
  
  if(NROW(cpsr_clinvar_variants) > 0){
    cpsr_clinvar_variants <- cpsr_clinvar_variants |>
      dplyr::select(-c(CPSR_CLASSIFICATION,CPSR_PATHOGENICITY_SCORE,
                       CPSR_CLASSIFICATION_DOC,CPSR_CLASSIFICATION_CODE)) |>
    dplyr::select(SYMBOL, CLINVAR_CLASSIFICATION,
                  dplyr::everything())
  }
  
    
  cpsr_other_variants <- class543_germline_variants |>
    dplyr::filter(CPSR_CLASSIFICATION_SOURCE == "Other")
  
  if(NROW(cpsr_other_variants) > 0){
    cpsr_other_variants <- cpsr_other_variants |>
      dplyr::select(-c(CLINVAR, 
                       CLINVAR_REVIEW_STATUS_STARS, 
                       CLINVAR_VARIANT_ORIGIN, 
                       CLINVAR_CLASSIFICATION, 
                       CLINVAR_CONFLICTED, 
                       CLINVAR_PHENOTYPE)) |>
      dplyr::select(SYMBOL, CPSR_CLASSIFICATION,
                    dplyr::everything())
  }
  
}

cpsr_biomarker_eitems <- data.frame()

for(n in c('diagnostic','prognostic','predictive','predisposing')){
  if(NROW(cpsr_rep[['content']][['snv_indel']][['clin_eitem']][[n]]) > 0){
    cpsr_biomarker_eitems <- cpsr_biomarker_eitems |>
      dplyr::bind_rows(
        cpsr_rep[['content']][['snv_indel']][['clin_eitem']][[n]]
      ) |>
      dplyr::select(SYMBOL, GENE_NAME, CANCER_TYPE, EVIDENCE_TYPE,
                    EVIDENCE_LEVEL, dplyr::everything())
  }
}

n_biomarkers <- cpsr_biomarker_eitems |> NROW()

tag_gnomad <- cpsr_rep[['metadata']][['config']][['popgen']][['vcftag_gnomad']]
desc_gnomad <- cpsr_rep[['metadata']][['config']][['popgen']][['popdesc_gnomad']]
formula_gnomad <- as.formula(paste0("~",rlang::sym(tag_gnomad)))


```

<h3>CPSR - sample metadata</h3> 

* Sample name &nbsp;:&nbsp; <span style="background-color:#000000; color:white"><b>&nbsp;`r cpsr_rep[['metadata']][['sample_name']]`&nbsp;</b></span>
<br>

<h3>CPSR - virtual gene panel</h3> 

Cancer predisposition geneset subject to analysis/screening in the CPSR report:

   *  __[`r cpsr_rep[['metadata']][['gene_panel']][['name']]`](`r cpsr_rep[['metadata']][['gene_panel']][['url']]`)__
      * Diagnostic-grade genes (applicable to Genomics England panels only): __`r cpsr_rep[['metadata']][['config']][['diagnostic_grade_only']]`__
   *  `r cpsr_rep[['metadata']][['gene_panel']][['confidence']]`
   
```{r cpsr_gene_selection, echo = F, eval = T}
tiles_html <- pcgrr::virtual_panel_display_html(gene_df = cpsr_rep[['metadata']][['gene_panel']][['genes']])
```

`r tiles_html`

<br><br>

<h3>CPSR - ClinVar variants</h3>

*  For sample __`r cpsr_rep[['metadata']][['sample_name']]`__, a total of n = __`r NROW(cpsr_clinvar_variants)`__ **germline variants** are registered with a <i>Pathogenic/Likely pathogenic/VUS</i> clinical significance in ClinVar.


```{r cpsr_clinvar_findings, echo=F, results = 'asis', eval = NROW(cpsr_clinvar_variants) > 0}

cat('\n')
htmltools::br()

if(NROW(cpsr_clinvar_variants) > 2000){
  cat('<b>NOTE - only considering top 2000 variants (due to limitations with client-side tables) </b><br>',sep="\n")
  cat('<br>')
  cpsr_clinvar_variants <- 
    head(cpsr_clinvar_variants, 2000)
}

cpsr_clinvar_variants_ctalk <- crosstalk::SharedData$new(cpsr_clinvar_variants)
crosstalk::bscols(
  list(
    crosstalk::filter_select("CLINVAR_CLASSIFICATION", "Clinical significance", cpsr_clinvar_variants_ctalk, ~CLINVAR_CLASSIFICATION),
    crosstalk::filter_select("CONSEQUENCE", "Consequence", cpsr_clinvar_variants_ctalk, ~CONSEQUENCE),
    crosstalk::filter_checkbox("GENOTYPE", "Genotype", cpsr_clinvar_variants_ctalk, ~GENOTYPE),
    crosstalk::filter_select("SYMBOL", "Gene", cpsr_clinvar_variants_ctalk, ~SYMBOL),
    crosstalk::filter_slider("GERP_SCORE","Genomic conservation score (GERP)",cpsr_clinvar_variants_ctalk, ~GERP_SCORE,
                             min = -12.3, max = 6.17, ticks = T),
    crosstalk::filter_slider(tag_gnomad, paste0("MAF gnomAD (",desc_gnomad,")"), cpsr_clinvar_variants_ctalk, formula_gnomad, sep="",ticks = F)
  ),
  list(
    crosstalk::filter_select("miRNA_TARGET_HIT", "miRNA target gain/loss", cpsr_clinvar_variants_ctalk, ~miRNA_TARGET_HIT),
    crosstalk::filter_select("TF_BINDING_SITE_VARIANT", "TF binding site alteration", cpsr_clinvar_variants_ctalk, ~TF_BINDING_SITE_VARIANT),
    crosstalk::filter_select("CLINVAR_PHENOTYPE", "ClinVar phenotype(s)", cpsr_clinvar_variants_ctalk, ~CLINVAR_PHENOTYPE),
    crosstalk::filter_slider("CLINVAR_REVIEW_STATUS_STARS", "ClinVar review status stars", cpsr_clinvar_variants_ctalk, ~CLINVAR_REVIEW_STATUS_STARS, min = 0, max = 4, step = 1, ticks = T),
    crosstalk::filter_select("CLINVAR_CONFLICTED", "ClinVar conflicting interpretations", cpsr_clinvar_variants_ctalk, ~CLINVAR_CONFLICTED)
  )
)

htmltools::br()
htmltools::br()

DT::datatable(
  cpsr_clinvar_variants_ctalk, 
  escape = F,
  extensions = c("Buttons","Responsive"),
  options = list(
    pageLength = 8,
    scrollCollapse = T,
    buttons = c('csv','excel'),
    dom = 'Bfrtip'
  )
) |>
  DT::formatStyle(
    columns = c("SYMBOL"),
    valueColumns = c("CLINVAR_CLASSIFICATION"), 
    color = "white", 
    backgroundColor = 
      DT::styleEqual(
        c("Pathogenic", "Likely_Pathogenic","VUS"), 
        c("#9E0142","#D53E4F","#000000")
      )
  )

htmltools::br()
htmltools::br()

```

<h3>CPSR - Non-ClinVar variants</h3>

*  For sample __`r cpsr_rep[['metadata']][['sample_name']]`__, a total of n = __`r NROW(cpsr_other_variants)`__ non-ClinVar **germline variants**
are classified with a <i>Pathogenic/Likely pathogenic/VUS</i> clinical significance according to CPSR.


```{r cpsr_other_findings, echo=F, results = 'asis', eval = NROW(cpsr_other_variants) > 0}

cat('\n')
htmltools::br()

if(NROW(cpsr_other_variants) > 2000){
  cat('<b>NOTE - only considering top 2000 variants (due to limitations with client-side tables) </b><br>',sep="\n")
  cat('<br>')
  cpsr_other_variants <- 
    head(cpsr_other_variants, 2000)
}

cpsr_other_variants_ctalk <- crosstalk::SharedData$new(cpsr_other_variants)
crosstalk::bscols(
  list(
    crosstalk::filter_select("CPSR_CLASSIFICATION", "Clinical significance", cpsr_other_variants_ctalk, ~CPSR_CLASSIFICATION),
    crosstalk::filter_select("CONSEQUENCE", "Consequence", cpsr_other_variants_ctalk, ~CONSEQUENCE),
    crosstalk::filter_checkbox("GENOTYPE", "Genotype", cpsr_other_variants_ctalk, ~GENOTYPE),
    crosstalk::filter_select("SYMBOL", "Gene", cpsr_other_variants_ctalk, ~SYMBOL),
    crosstalk::filter_slider("GERP_SCORE","Genomic conservation score (GERP)",cpsr_other_variants_ctalk, ~GERP_SCORE,
                             min = -12.3, max = 6.17, ticks = T)
  ),
  list(
    crosstalk::filter_select("miRNA_TARGET_HIT", "miRNA target gain/loss", cpsr_other_variants_ctalk, ~miRNA_TARGET_HIT),
    crosstalk::filter_select("TF_BINDING_SITE_VARIANT", "TF binding site alteration", cpsr_other_variants_ctalk, ~TF_BINDING_SITE_VARIANT),
    crosstalk::filter_select("CPSR_CLASSIFICATION_CODE", "CPSR classification (ACMG criteria codes)", cpsr_other_variants_ctalk, ~CPSR_CLASSIFICATION_CODE),
    crosstalk::filter_slider("CPSR_PATHOGENICITY_SCORE", "CPSR pathogenicity score", cpsr_other_variants_ctalk, ~CPSR_PATHOGENICITY_SCORE, step = 0.5, ticks = T),
    crosstalk::filter_slider(tag_gnomad, paste0("MAF gnomAD (",desc_gnomad,")"), cpsr_other_variants_ctalk, formula_gnomad, sep="",ticks = F)
  )
)
htmltools::br()
htmltools::br()

DT::datatable(
  cpsr_other_variants_ctalk, 
  escape = F,
  extensions = c("Buttons","Responsive"),
  options = list(
    pageLength = 8,
    scrollCollapse = T,
    buttons = c('csv','excel'),
    dom = 'Bfrtip'
  )
) |>
  DT::formatStyle(
    columns = c("SYMBOL"),
    valueColumns = c("CPSR_CLASSIFICATION"),
    color = "white", 
    backgroundColor = DT::styleEqual(
      c("Pathogenic", "Likely_Pathogenic","VUS"), 
      c("#9E0142","#D53E4F","#000000"))
  )

htmltools::br()
htmltools::br()

```

<h3>CPSR - Biomarkers</h3>

* CPSR detected __n = `r NROW(cpsr_biomarker_eitems)`__ biomarker evidence items that overlap **germline variants** in the input sample.

```{r cpsr_biomarker_note, echo=F, results = "asis", include = NROW(cpsr_biomarker_eitems) > 0}

cat('<b>NOTE:</b> Reported biomarkers in CIViC are mapped at different resolutions (i.e. filter <b>Biomarker mapping</b>). The accuracy of a match between variants in the tumor sample and the reported biomarkers will vary accordingly (highlighted by gene symbols with different color backgrounds):\n\n')

cat('<ul><li>Biomarker match at the <mark style="background-color:black; font-weight:bold; color:white">exact variant/codon level</mark></li>')
cat(paste0('<br><li>Biomarker match at the <mark style="background-color:', cpsr_rep[['metadata']][['color_palette']][['warning']][['values']][1],'; color:white; font-weight:bold">exon/gene level</mark></li></ul>\n'))

htmltools::br()
```


```{r cpsr_list_biomarker_eitems, echo=F, results = 'asis', eval = NROW(cpsr_biomarker_eitems) > 0}


cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()

DT::datatable(
  cpsr_biomarker_eitems, 
  escape=F, 
  extensions = 
    c("Buttons","Responsive"), 
  options = 
    list(buttons = c('csv','excel'),
         pageLength = 5,
         dom = 'Bfrtip')
  ) |>
  DT::formatStyle(
    'EVIDENCE_LEVEL', 
    backgroundColor = DT::styleEqual(
      c('A: Validated', 
        'A: FDA/NCCN/ELN guidelines', 
        'B: Clinical evidence', 
        'B1: Clinical evidence: late trials', 
        'B2: Clinical evidence: early trials', 
        'C: Case study', 
        'D: Preclinical evidence', 
        'E: Indirect evidence'), 
      c("#009E73","#009E73","#56B4E9", 
        "#56B4E9","#56B4E9","#0072B2",
        "#E69F00", "#F0E442"))) |>
  DT::formatStyle(
    color="white", "SYMBOL", 
    "BIOMARKER_MAPPING", 
    fontWeight = 'bold', `text-align` = 'center',
    backgroundColor = DT::styleEqual(
      c('exact','codon','exon','gene'), c('#000','#000',                                      cpsr_rep[['metadata']][['color_palette']][['warning']][['values']][1], cpsr_rep[['metadata']][['color_palette']][['warning']][['values']][1])))



```

<br><br>

