### Tumor mutational burden (TMB)

For estimation of TMB, PCGR employs two different approaches/algorithms ( __all_coding__, and __nonsyn__, see details outlined in the [Documentation](#Documentation) below).

TMB algorithm chosen by user: __`r pcg_report[['metadata']][['config']][['tmb']][['algorithm']]`__
Size of targeted coding region: __`r pcg_report[['metadata']][['config']][["assay_props"]][['target_size_mb']]` Mb__

Estimated mutational burden: __`r pcg_report[['content']][['tmb']][['v_stat']][['tmb_estimate']]`__ __mutations/Mb__ 

<br><br>

```{r tumor_tmb_tcga_header, echo = F, results='asis', eval=T}
cat('<h4>TMB reference distributions - TCGA</h4>',sep="\n")
```
<br>
The plot below indicates how the mutational burden estimated for the query tumor sample (red dotted line) compares with the distributions observed for tumor samples in The Cancer Genome Atlas (TCGA). The grey area indicates the upper TMB tertile as defined by the user. Please note the following characteristics of the TCGA dataset presented here, which must be taken into account during TMB interpretation of the query sample:

* The TCGA tumor samples are sequenced with a mean coverage of approximately 100X
* The TCGA somatic mutation calls are based on a consensus among variant callers (each variant is supported by a minimum of two variant calling algorithms)
* The TCGA somatic mutation calls are based on paired tumor-control sequencing (tumor-only sequencing may produce higher numbers due to more noise)

<br><br>
```{r tumor_tmb_tcga_plot, echo = F, fig.width=12, fig.height=9, eval=T}
tmb_plot <- pcgrr::plot_tmb_primary_site_tcga(pcg_report[['content']][['tmb']][['tcga_tmb']], 
                                              p_site = pcg_report[['metadata']][['config']][["t_props"]][['tumor_type']], 
                                              tmb_estimate = pcg_report[['content']][['tmb']][['v_stat']][['tmb_estimate']],
                                              algorithm = pcg_report[['content']][['tmb']][['algorithm']])
tmb_plot
htmltools::br()
```

<br>

