## Settings

```{r on_off_settings, echo=F}
signatures_estimation <- 'ON'
msi_prediction <- 'ON'
mutational_burden <- 'ON'
if(pcg_report$content$msi$eval == F){
  msi_prediction <- 'OFF'
}
if(pcg_report$content$m_signature_mp$eval == F){
  signatures_estimation <- 'OFF'
}
if(pcg_report$content$tmb$eval == F){
  mutational_burden <- 'OFF'
}

gencode_versions <- unlist(stringr::str_split_fixed(
  stringr::str_replace(
    pcg_report$metadata$pcgr_db_release$gencode$version,"release ",""),
  "/", 2))[1,]
  
gencode_version <- gencode_versions[2]
if(pcg_report[['metadata']][['genome_assembly']] == "grch38"){
  gencode_version <- gencode_versions[1]
}

transcript_set_gencode <- paste0("GENCODE - basic set (v", gencode_version,")")
if(pcg_report[['metadata']][['config']][['other']][['vep_gencode_all']] == T){
  transcript_set_gencode <- paste0("GENCODE - all transcripts (v", gencode_version,")")
}

```

### Metadata - sample and sequencing assay

* Sample name &nbsp;:&nbsp; <span style="background-color:#000000; color:white"><b>&nbsp;`r pcg_report[['metadata']][['sample_name']]`&nbsp;</b></span>
* Tumor primary site:  __`r pcg_report[['metadata']][['config']][["t_props"]][['tumor_type']]`__
* Sequencing mode input (VCF): __`r pcg_report[['metadata']][['config']][["assay_props"]][['mode']]`__
* Sequencing type input (VCF): __`r pcg_report[['metadata']][['config']][["assay_props"]][['type']]`__
* Coding target size (VCF): __`r pcg_report[['metadata']][['config']][["assay_props"]][['target_size_mb']]` Mb__

### Report configuration

The report is generated with _[PCGR](https://github.com/sigven/pcgr)_ version `r pcg_report[['metadata']][['pcgr_version']]`, using the following key settings:

* Minimum sequencing depth (DP) tumor (SNV + InDels): __`r pcg_report[['metadata']][['config']][['allelic_support']][['tumor_dp_min']]`__
* Minimum allelic fraction (AF) tumor (SNV + InDels): __`r pcg_report[['metadata']][['config']][['allelic_support']][['tumor_af_min']]`__
* Minimum sequencing depth (DP) control (SNV + InDels): __`r pcg_report[['metadata']][['config']][['allelic_support']][['control_dp_min']]`__
* Maximum allelic fraction (AF) control (SNV + InDels): __`r pcg_report[['metadata']][['config']][['allelic_support']][['control_af_max']]`__
* Tier system (VCF): __pcgr_acmg__
* Show noncoding variants: __`r pcg_report$metadata$config$other$list_noncoding`__
* MSI prediction: __`r msi_prediction`__
* Mutational burden estimation: __`r mutational_burden`__
  * TMB algorithm: <b> __`r pcg_report[['metadata']][['config']][['tmb']][['algorithm']]`__ </b>
* Mutational signatures estimation: __`r signatures_estimation`__
  * Minimum number of mutations required: __`r pcg_report[['metadata']][['config']][['msigs']][['mutation_limit']]`__
  * All reference signatures: <b> __`r pcg_report[['metadata']][['config']][['msigs']][['all_reference_signatures']]`__ </b>
  * Inclusion of artefact signatures: <b> __`r pcg_report[['metadata']][['config']][['msigs']][['include_artefact_signatures']]`__ </b>
* Report theme (Bootstrap): __`r pcg_report[['metadata']][['config']][['visual']][['report_theme']]`__
* Variant Effect Predictor (VEP) settings:
  * Transcript set: __`r transcript_set_gencode`__
  * Transcript pick order: __`r pcg_report[['metadata']][['config']][['other']][['vep_pick_order']]`__
  * Regulatory regions annotation: __`r pcg_report[['metadata']][['config']][['other']][['vep_regulatory']]`__



