
### Tier 2 - Variants of potential clinical significance {.tabset}

- Tier 2 considers evidence items of _i)_ strong evidence levels (A & B) in other tumor types, and _ii)_ weak evidence levels (C, D & E) in the query tumor type (_`r pcg_report[['metadata']][['config']][["t_props"]][['tumor_type']]`_). Using the [database for clinical interpretations of variants in cancer (CIViC) ](https://civic.genome.wustl.edu/#/home) and [Cancer Biomarkers database](https://www.cancergenomeinterpreter.org/biomarkers), a total of __`r NROW(pcg_report[['content']][['snv_indel']][["disp"]][['tier2']])`__ unique, somatic variants were found in the tumor sample:
    * Tier 2 - Predictive/Therapeutic: __`r NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['other_ttype']][['predictive']][['A_B']]) + NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][['predictive']][['C_D_E']])`__ evidence items
    * Tier 2 - Prognostic: __`r NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['other_ttype']][['prognostic']][['A_B']]) + NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][['prognostic']][['C_D_E']])`__ evidence items
    * Tier 2 - Diagnostic: __`r NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['other_ttype']][['diagnostic']][['A_B']]) + NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][['diagnostic']][['C_D_E']])`__ evidence items

<br><br>


```{r biomarker_note2, echo=F, results = "asis", include = pcg_report$content$snv_indel$v_stat$n_tier2 > 0}

cat('<b>NOTE:</b> Reported biomarkers in CIViC/CGI are mapped at different resolutions (i.e. filter <b>Biomarker mapping</b>). The accuracy of a match between variants in the tumor sample and the reported biomarkers will vary accordingly (highlighted by gene symbols with different color backgrounds):\n\n')

cat('<ul><li>Biomarker match at the <mark style="background-color:black; font-weight:bold; color:white">exact variant/codon level</mark></li>')
cat(paste0('<br><li>Biomarker match at the <mark style="background-color:', pcg_report[['metadata']][['color_palette']][['warning']][['values']][1],'; color:white; font-weight:bold">exon/gene level</mark></li></ul>\n'))

htmltools::br()
```


```{r prepare_tier2_data, echo=F, results='asis'}

show_tier2_filters <- list()
missing_tier2_items <- list()
eitems <- list()
for(type in c('diagnostic','prognostic','predictive')){
  show_tier2_filters[[type]] <- F
  missing_tier2_items[[type]] <- T
  eitems[[type]] <- data.frame()
  if(NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['other_ttype']][[type]][['A_B']]) > 0 | NROW(pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][[type]][['C_D_E']]) > 0){
    eitems[[type]] <- dplyr::bind_rows(pcg_report[['content']][['snv_indel']][['clin_eitem']][['other_ttype']][[type]][['A_B']], pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][[type]][['C_D_E']]) %>%
      dplyr::arrange(desc(RATING))
    show_tier2_filters[[type]] <- T
    missing_tier2_items[[type]] <- F
  }
}

```


```{r active_tab_snv_predictive_tier2, echo = F, results = "asis"}
if(missing_tier2_items[['predictive']] == F | 
   (missing_tier2_items[['diagnostic']] == T & 
    missing_tier2_items[['prognostic']] == T)){
  cat("")
  cat("#### Predictive biomarkers {.active}")
  cat("")
}else{
  cat("")
  cat("#### Predictive biomarkers")
  cat("")
}
```

<br>

```{r tier2_predictive, echo=F, results = 'asis', eval = show_tier2_filters[['predictive']]}

variants_tier2_predictive_shared <- crosstalk::SharedData$new(eitems[['predictive']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_tier2_predictive_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_tier2_predictive_shared, ~CLINICAL_SIGNIFICANCE),
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_tier2_predictive_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_slider("RATING", "Rating", variants_tier2_predictive_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)
  ),
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_tier2_predictive_shared, ~CONSEQUENCE),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_tier2_predictive_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_select("THERAPEUTIC_CONTEXT", "Therapeutic context", variants_tier2_predictive_shared, ~THERAPEUTIC_CONTEXT)
  )
)

cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_tier2_predictive_shared,  escape=F,extensions=c("Buttons","Responsive"), 
              options=list(buttons = c('csv','excel'), scrollCollapse = T, dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated','A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence','B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials','C: Case study',
                                                     'D: Preclinical evidence','E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) %>%
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))


htmltools::br()
htmltools::br()

```

```{r tier2_predictive_missing, echo=F, results = 'asis', eval = missing_tier2_items[['predictive']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br>


```{r active_tab_snv_prognostic_tier2, echo = F, results = "asis"}
if(missing_tier2_items[['prognostic']] == F & 
   missing_tier2_items[['predictive']] == T){
  cat("")
  cat("#### Prognostic biomarkers {.active}")
  cat("")
}else{
  cat("")
  cat("#### Prognostic biomarkers")
  cat("")
}
```

<br>

```{r tier2_prognostic, echo=F, results='asis', eval = show_tier2_filters[['prognostic']]}

variants_tier2_prognostic_shared <- crosstalk::SharedData$new(eitems[['prognostic']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_tier2_prognostic_shared, ~CONSEQUENCE),
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_tier2_prognostic_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_tier2_prognostic_shared, ~CLINICAL_SIGNIFICANCE)
  ),
  list(
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_tier2_prognostic_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_tier2_prognostic_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_slider("RATING", "Rating", variants_tier2_prognostic_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)
  )
)
cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_tier2_prognostic_shared, escape=F,extensions=c("Buttons","Responsive"), 
              options=list(buttons = c('csv','excel'), scrollCollapse = T, dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated','A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence','B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials','C: Case study',
                                                     'D: Preclinical evidence','E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) %>%
   DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))


htmltools::br()
htmltools::br()

```

```{r tier2_prognostic_missing, echo=F, results = 'asis', eval = missing_tier2_items[['prognostic']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br>


```{r active_tab_snv_diagnostic_tier2, echo = F, results = "asis"}
if(missing_tier2_items[['diagnostic']] == F & 
   missing_tier2_items[['predictive']] == T & 
   missing_tier2_items[['prognostic']] == T){
  cat("")
  cat("#### Diagnostic biomarkers {.active}")
  cat("")
}else{
  cat("")
  cat("#### Diagnostic biomarkers")
  cat("")
}
```

<br>

```{r tier2_diagnostic, echo=F, results='asis', eval = show_tier2_filters[['diagnostic']]}

variants_tier2_diagnostic_shared <- crosstalk::SharedData$new(eitems[['diagnostic']])
crosstalk::bscols(
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_tier2_diagnostic_shared, ~CONSEQUENCE),
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_tier2_diagnostic_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical association", variants_tier2_diagnostic_shared, ~CLINICAL_SIGNIFICANCE)
  ),
  list(
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_tier2_diagnostic_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_tier2_diagnostic_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_slider("RATING", "Rating", variants_tier2_diagnostic_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)
  )
)
cat('The table below lists all variant-evidence item associations:',sep='\n')
htmltools::br()
htmltools::br()

DT::datatable(variants_tier2_diagnostic_shared, 
              escape=F,extensions=c("Buttons","Responsive"), 
              options=list(buttons = c('csv','excel'), scrollCollapse = T, dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                  backgroundColor = DT::styleEqual(c('A: Validated',
                                                     'A: FDA/NCCN/ELN guidelines',
                                                     'B: Clinical evidence',
                                                     'B1: Clinical evidence: late trials',
                                                     'B2: Clinical evidence: early trials',
                                                     'C: Case study',
                                                     'D: Preclinical evidence',
                                                     'E: Indirect evidence'), 
                                                   c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) %>%
   DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))


htmltools::br()
htmltools::br()

```

```{r tier2_diagnostic_missing, echo=F, results = 'asis', eval = missing_tier2_items[['diagnostic']]}
cat('<i>No variant-evidence item associations found.</i>',sep='\n')
cat('\n')
```

<br><br>


