### Complete biomarker set 

```{r prepare_bm_data, eval = T}

all_evidence_items <- pcg_report$content$snv_indel$clin_eitem$any_ttype$diagnostic$any %>%
     dplyr::bind_rows(pcg_report$content$snv_indel$clin_eitem$any_ttype$prognostic$any) %>%
     dplyr::bind_rows(pcg_report$content$snv_indel$clin_eitem$any_ttype$predictive$any)

n_all <- NROW(all_evidence_items)

n_entries_diagnostic <- 0
if(NROW(all_evidence_items) > 0){
  n_entries_diagnostic <- all_evidence_items %>%
    dplyr::filter(EVIDENCE_TYPE == "Diagnostic") %>%
    nrow()
}
n_entries_prognostic <- 0
if(NROW(all_evidence_items) > 0){
  n_entries_prognostic <- all_evidence_items %>%
    dplyr::filter(EVIDENCE_TYPE == "Prognostic") %>%
    nrow()
}
n_entries_predictive <- 0
if(NROW(all_evidence_items) > 0){
  n_entries_predictive <- all_evidence_items %>%
    dplyr::filter(EVIDENCE_TYPE == "Predictive") %>%
    nrow()
}

```


<br>
A list of __ALL__ associations between variants in the tumor sample and known biomarkers are shown in this section, i.e. also listing biomarkers that are not assigned to TIER 1/TIER 2.
<br>

```{r biomarker_note_5, echo=F, results = "asis", include = n_all > 0}

cat('<b>NOTE:</b> Reported biomarkers in CIViC/CGI are mapped at different resolutions (i.e. filter <b>Biomarker mapping</b>). The accuracy of a match between variants in the tumor sample and the reported biomarkers will vary accordingly (highlighted by gene symbols with different color backgrounds):\n\n')

cat('<ul><li>Biomarker match at the <mark style="background-color:black; font-weight:bold; color:white">exact variant/codon level</mark></li>')
cat(paste0('<br><li>Biomarker match at the <mark style="background-color:', pcg_report[['metadata']][['color_palette']][['warning']][['values']][1],'; color:white; font-weight:bold">exon/gene level</mark></li></ul>\n'))

htmltools::br()
```



```{r table_browse_biomarkers, echo=F, results = "asis", eval = n_all > 0}

variants_any <- dplyr::bind_rows(
  pcg_report[['content']][['snv_indel']][['clin_eitem']][['any_ttype']][['diagnostic']][['any']],
  pcg_report[['content']][['snv_indel']][['clin_eitem']][['any_ttype']][['predictive']][['any']],
  pcg_report[['content']][['snv_indel']][['clin_eitem']][['any_ttype']][['prognostic']][['any']])

variants_all_shared <- crosstalk::SharedData$new(variants_any)
crosstalk::bscols(
  list(
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_all_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_all_shared, ~CLINICAL_SIGNIFICANCE),
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_all_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_slider("RATING", "Rating", variants_all_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)

  ),
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_all_shared, ~CONSEQUENCE),
    crosstalk::filter_select("EVIDENCE_TYPE", "Evidence type", variants_all_shared, ~EVIDENCE_TYPE),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_all_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_select("THERAPEUTIC_CONTEXT", "Therapeutic context", variants_all_shared, ~THERAPEUTIC_CONTEXT)
  )
)


```


```{r biomarker_missing_filters, echo=F, results = 'asis', eval = n_all == 0}
cat('\n*  <i> <font style="font-size: 100%"><b>No</b> biomarkers associated with variants in the tumor sample.</i></font>', sep='\n')
cat('\n')
```

<br>

```{r biomarker_table_all, eval = n_all > 0}

variants_all_shared %>%
  DT::datatable(escape = F, 
                extensions = c("Buttons","Responsive"), 
                options = list(pageLength = 8, 
                               buttons = c('csv','excel'), 
                               dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                 backgroundColor = DT::styleEqual(c('A: Validated',
                                                    'A: FDA/NCCN/ELN guidelines',
                                                    'B: Clinical evidence',
                                                    'B1: Clinical evidence: late trials',
                                                    'B2: Clinical evidence: early trials',
                                                    'C: Case study',
                                                    'D: Preclinical evidence',
                                                    'E: Indirect evidence'), 
                                                  c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) %>%
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))

```

