Tier 4 {data-navmenu="SNVs and InDels"}
================================================================== 

Row
-------------------------------------

```{r data_prep_tier4, include = F, eval = T}
n_tier4_snv <- 0
n_tier4_indel <- 0
n_tier4_block <- 0

if(NROW(pcg_report$content$snv_indel$callset$variant_display) > 0){
  n_tier4_snv <- 
    pcg_report$content$snv_indel$callset$variant_display |>
    dplyr::filter(.data$VARIANT_CLASS == "SNV" & 
                  !is.na(.data$TIER) & 
                  .data$TIER == 4) |> 
    nrow()
  n_tier4_indel <- 
    pcg_report$content$snv_indel$callset$variant_display |> 
    dplyr::filter((.data$VARIANT_CLASS == "insertion" | 
                     .data$VARIANT_CLASS == "deletion" | 
                     .data$VARIANT_CLASS == "indel") & 
                    !is.na(.data$TIER) &
                    .data$TIER == 4) |>
    nrow()
  n_tier4_block <- 
    pcg_report$content$snv_indel$callset$variant_display |> 
    dplyr::filter(.data$VARIANT_CLASS == "substitution" & 
                    !is.na(.data$TIER) &
                    .data$TIER == 4) |> 
    nrow()
}

```

### SNVs and InDels

```{r}
flexdashboard::valueBox(
  "TIER 4", caption = NULL, color = '#000000', icon = NULL)
```


### Total variants

```{r}
flexdashboard::valueBox(
  pcg_report$content$snv_indel$vstats$n_tier4, 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```

### SNVs

```{r}
flexdashboard::valueBox(
  as.character(n_tier4_snv), 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```

### InDels

```{r}
flexdashboard::valueBox(
  as.character(n_tier4_indel), 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```

### Block substitutions

```{r}
flexdashboard::valueBox(
  as.character(n_tier4_block), 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```



Row
-----------------------------------------------------

### Tier 4 - variant filters {data-padding=15 data-width=250}

<br>

```{r table_browse_tier4, echo=F, results = "asis", eval = pcg_report$content$snv_indel$vstats$n_tier4 > 0}

tier4_vars <- 
  pcg_report$content$snv_indel$callset$variant_display |>
  dplyr::filter(
    !is.na(.data$TIER) & 
      .data$TIER == 4) |>
  dplyr::select(
    dplyr::any_of(
      pcgrr::dt_display$tier4))
  
if(pcg_report$content$snv_indel$vstats$n_tier4 > 2000){
  cat('<b>NOTE - only considering top 2000 variants (ranked according to oncogenicity and gene-cancer association score) </b><br>',sep="\n")
  cat('<br>')
  tier4_vars <- 
    head(tier4_vars, 2000)
}
htmltools::br()

cat('Variants in TIER 4 (right panel) can be interactively explored according to various criteria\n:')

variants_tier4_shared <- crosstalk::SharedData$new(tier4_vars)

crosstalk::bscols(
  list(
    crosstalk::filter_select(
      "CONSEQUENCE", "Consequence", 
      variants_tier4_shared, ~CONSEQUENCE),
    crosstalk::filter_select(
      "SYMBOL", "Gene", 
      variants_tier4_shared, ~SYMBOL)
  )
)


```


```{r tier4_missing_filters, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$vstats$n_tier4 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> other protein-coding variants found.</font></i>', sep = '\n')
cat('\n')
```


### Tier 4 - variant datatable {data-padding=15}

<br>

```{r table_tier4, eval = pcg_report$content$snv_indel$vstats$n_tier4 > 0}

DT::datatable(
  variants_tier4_shared, 
  escape = F,
  extensions = c("Buttons","Responsive"), 
  options = list(
    buttons = c('csv','excel'), 
    scrollCollapse=T, 
    scrollY = scrollY_flexdb_container,
    fixedColumns = T,
    fixedHeader = T,
    width = "100%",
    scrollX = "900px",
    columnDefs = list(list(
      width = '90px', targets = c(1,2,3,4,5,6))),
    dom = 'Bfrtip')) |>
  DT::formatStyle(
    "SYMBOL",
    "GLOBAL_ASSOC_RANK", 
    color = "white", 
    backgroundColor = 
      DT::styleInterval(
        pcgrr::color_palette$cancer_assoc$breaks, 
        pcgrr::color_palette$cancer_assoc$values))

```

```{r tier4_missing_data, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$vstats$n_tier4 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> other protein-coding variants found.</font></i>', sep = '\n')
cat('\n')
```

