Tier 1 {data-navmenu="SNVs and InDels"}
==================================================================

Row
-------------------------------------

### SNVs and InDels

```{r}
flexdashboard::valueBox("TIER 1", caption = NULL, color = '#000000', icon = NULL)
```

### Biomarker genes

```{r}

tier1_evidence_items <- pcg_report$content$snv_indel$clin_eitem$specific_ttype$diagnostic$A_B %>%
     dplyr::bind_rows(pcg_report$content$snv_indel$clin_eitem$specific_ttype$prognostic$A_B) %>%
     dplyr::bind_rows(pcg_report$content$snv_indel$clin_eitem$specific_ttype$predictive$A_B)

flexdashboard::valueBox(length(unique(tier1_evidence_items$SYMBOL)), 
                        color = pcg_report$metadata$color_value_box)

```

### Biomarker variants

```{r}
flexdashboard::valueBox(pcg_report$content$snv_indel$v_stat$n_tier1, 
                        color = pcg_report$metadata$color_value_box, icon = "fa-dna")
```


### Diagnostic evidence items

```{r}
entries_diagnostic <- NROW(pcg_report$content$snv_indel$clin_eitem$specific_ttype$diagnostic$A_B)
flexdashboard::valueBox(entries_diagnostic, color = ifelse(entries_diagnostic > 0,
                        pcg_report[['metadata']][['color_palette']][['success']][['values']][1],
                        pcg_report$metadata$color_none), icon = "fa-file-prescription")
```

### Prognostic evidence items

```{r}
entries_prognostic <- NROW(pcg_report$content$snv_indel$clin_eitem$specific_ttype$prognostic$A_B)
flexdashboard::valueBox(entries_prognostic, color = ifelse(entries_prognostic > 0,
                        pcg_report[['metadata']][['color_palette']][['success']][['values']][1],
                        pcg_report$metadata$color_none), icon = "fa-file-prescription")
```

### Predictive evidence items

```{r}
entries_predictive <- NROW(pcg_report$content$snv_indel$clin_eitem$specific_ttype$predictive$A_B)
flexdashboard::valueBox(entries_predictive, color = ifelse(entries_predictive > 0,
                        pcg_report[['metadata']][['color_palette']][['success']][['values']][1],
                        pcg_report$metadata$color_none), icon = "fa-file-prescription")
```


Row
-----------------------------------------------------

### Tier 1 variant evidence items - filters {data-padding=15 data-width=300}

<br>

```{r table_browse_tier1, echo=F, results = "asis", eval = pcg_report$content$snv_indel$v_stat$n_tier1 > 0}

cat('\nEvidence items associated with variants in tier 1 (right panel) can be filtered according to various criteria:\n')

variants_tier1 <- dplyr::bind_rows(
  pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][['diagnostic']][['A_B']],
  pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][['predictive']][['A_B']],
  pcg_report[['content']][['snv_indel']][['clin_eitem']][['specific_ttype']][['prognostic']][['A_B']])

variants_tier1_shared <- crosstalk::SharedData$new(variants_tier1)
crosstalk::bscols(
  list(
    crosstalk::filter_select("CANCER_TYPE", "Cancer type", variants_tier1_shared, ~CANCER_TYPE),
    crosstalk::filter_select("CLINICAL_SIGNIFICANCE", "Clinical significance", variants_tier1_shared, ~CLINICAL_SIGNIFICANCE),
    crosstalk::filter_select("EVIDENCE_LEVEL", "Evidence level", variants_tier1_shared, ~EVIDENCE_LEVEL),
    crosstalk::filter_slider("RATING", "Rating", variants_tier1_shared, ~RATING, min = 0, max = 5, step = 1, ticks = T)

  ),
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_tier1_shared, ~CONSEQUENCE),
    crosstalk::filter_select("EVIDENCE_TYPE", "Evidence type", variants_tier1_shared, ~EVIDENCE_TYPE),
    crosstalk::filter_select("BIOMARKER_MAPPING", "Biomarker mapping", variants_tier1_shared, ~BIOMARKER_MAPPING),
    crosstalk::filter_select("THERAPEUTIC_CONTEXT", "Therapeutic context", variants_tier1_shared, ~THERAPEUTIC_CONTEXT)
  )
)


```


```{r biomarker_note1, echo=F, results = "asis", include = pcg_report$content$snv_indel$v_stat$n_tier1 > 0}

cat('<b>NOTE:</b> Reported biomarkers in CIViC/CGI are mapped at different resolutions (i.e. filter <b>Biomarker mapping</b>). The accuracy of a match between variants in the tumor sample and the reported biomarkers will vary accordingly (highlighted by gene symbols with different color backgrounds):\n\n')

cat('<ul><li>Biomarker match at the <mark style="background-color:black; font-weight:bold; color:white">exact variant/codon level</mark></li>')
cat(paste0('<br><li>Biomarker match at the <mark style="background-color:', pcg_report[['metadata']][['color_palette']][['warning']][['values']][1],'; color:white; font-weight:bold">exon/gene level</mark></li></ul>\n'))

htmltools::br()
htmltools::br()
```


```{r tier1_missing_filters, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$v_stat$n_tier1 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> variants of strong clinical significance found.</i></font>', sep='\n')
cat('\n')
```


### Tier 1  - variant evidence items {data-padding=15}

<br>

```{r table_tier1, eval = pcg_report$content$snv_indel$v_stat$n_tier1 > 0}

variants_tier1_shared %>%
  DT::datatable(escape = F, 
               extensions = c("Buttons","Responsive"), 
                options = list(pageLength = 8, 
                               scrollY = scrollY_flexdb_container, 
                               buttons = c('csv','excel'), 
                               #autoWidth = T, 
                 fixedColumns = T,
                 fixedHeader = T,
                 width = "100%",
                 scrollX = "900px",
                  columnDefs = list(list(width = '90px', targets = c(1,2,3,4,5,6))),
                               dom = 'Bfrtip')) %>%
  DT::formatStyle('EVIDENCE_LEVEL', 
                 backgroundColor = DT::styleEqual(c('A: Validated',
                                                    'A: FDA/NCCN/ELN guidelines',
                                                    'B: Clinical evidence',
                                                    'B1: Clinical evidence: late trials',
                                                    'B2: Clinical evidence: early trials',
                                                    'C: Case study',
                                                    'D: Preclinical evidence',
                                                    'E: Indirect evidence'), 
                                                  c("#009E73","#009E73","#56B4E9", "#56B4E9","#56B4E9","#0072B2","#E69F00", "#F0E442"))) %>%
  DT::formatStyle(color="white", "SYMBOL", "BIOMARKER_MAPPING", fontWeight = 'bold', `text-align` = 'center',
                  backgroundColor = DT::styleEqual(c('exact','codon','exon','gene'), 
                                                   c('#000','#000',pcg_report[['metadata']][['color_palette']][['warning']][['values']][1], pcg_report[['metadata']][['color_palette']][['warning']][['values']][1])))

```


```{r tier1_missing_data, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$v_stat$n_tier1 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> variants of strong clinical significance found.</i></font>',sep='\n')
cat('\n')
```

