Noncoding {data-navmenu="SNVs and InDels"}
================================================================== 

Row
-------------------------------------

```{r data_prep_noncoding, include = F, eval = T}
n_noncoding_snv <- 0
n_noncoding_indel <- 0
n_noncoding_block <- 0

if(NROW(pcg_report$content$snv_indel$variant_set$tsv) > 0){
  n_noncoding_snv <- pcg_report$content$snv_indel$variant_set$tsv |>
    dplyr::filter(VARIANT_CLASS == "SNV" & TIER == "NONCODING") |> 
    nrow()
  n_noncoding_indel <- pcg_report$content$snv_indel$variant_set$tsv |> 
    dplyr::filter((VARIANT_CLASS == "insertion" | 
                     VARIANT_CLASS == "deletion" | 
                     VARIANT_CLASS == "indel") & TIER == "NONCODING") |>
    nrow()
  n_noncoding_block <-  pcg_report$content$snv_indel$variant_set$tsv |> 
    dplyr::filter(VARIANT_CLASS == "substitution" & TIER == "NONCODING") |> 
    nrow()
}

```


### SNVs and InDels

```{r}
flexdashboard::valueBox("Noncoding", caption = NULL, color = '#000000', icon = NULL)
```


### Total variants

```{r}
flexdashboard::valueBox(pcg_report$content$snv_indel$v_stat$n_noncoding, color = pcg_report$metadata$color_value_box, icon = "fa-dna")
```

### SNVs

```{r}
flexdashboard::valueBox(as.character(n_noncoding_snv), color = pcg_report$metadata$color_value_box, icon = "fa-dna")
```

### InDels

```{r}
flexdashboard::valueBox(as.character(n_noncoding_indel), color = pcg_report$metadata$color_value_box, icon = "fa-dna")
```

### Block substitutions

```{r}
flexdashboard::valueBox(as.character(n_noncoding_block), color = pcg_report$metadata$color_value_box, icon = "fa-dna")
```

Row
-----------------------------------------------------

### Noncoding - variant filters {data-padding=15 data-width=250}

<br>

```{r table_browse_noncoding, echo=F, results = "asis", eval = pcg_report$content$snv_indel$v_stat$n_noncoding > 0}

if(NROW(pcg_report[['content']][['snv_indel']][["disp"]][['noncoding']]) > 2000){
  cat('<b>NOTE - only considering top 2000 variants (ranked according to Open Targets phenotype association score) </b><br>',sep="\n")
  cat('<br>')
  pcg_report[['content']][['snv_indel']][["disp"]][['noncoding']] <- 
    head(pcg_report[['content']][['snv_indel']][["disp"]][['noncoding']], 2000)
}
htmltools::br()


cat('Noncoding variants (right panel) can be interactively explored according to various criteria\n:')

variants_noncoding_shared <- 
  crosstalk::SharedData$new(pcg_report[['content']][['snv_indel']][["disp"]][['noncoding']])

crosstalk::bscols(
  list(
    crosstalk::filter_select("CONSEQUENCE", "Consequence", variants_noncoding_shared, ~CONSEQUENCE),
    crosstalk::filter_select("SYMBOL", "Gene", variants_noncoding_shared, ~SYMBOL)
  )
)

```


```{r noncoding_missing_filters, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$v_stat$n_noncoding == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> noncoding variants found.</font></i>', sep = '\n')
cat('\n')
```


### Noncoding  - variant datatable {data-padding=15}


```{r table_noncoding, eval = pcg_report$content$snv_indel$v_stat$n_noncoding > 0}

DT::datatable(variants_noncoding_shared, escape=F,extensions=c("Buttons","Responsive"), 
options=list(buttons = c('csv','excel'), 
             scrollCollapse = T,
             scrollY = scrollY_flexdb_container,
             #autoWidth = T, 
                 fixedColumns = T,
                 fixedHeader = T,
                 scrollX = "900px",
                  columnDefs = list(list(width = '90px', targets = c(1,2,3,4,5,6))),
             dom = 'Bfrtip')) |>
  DT::formatStyle("SYMBOL","OPENTARGETS_RANK", color = "white", backgroundColor = DT::styleInterval(pcg_report[['content']][['report_display_config']][['opentargets_rank']][['breaks']], pcg_report[['content']][['report_display_config']][['opentargets_rank']][['colors']]))
  
```

```{r noncoding_missing_data, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$v_stat$n_noncoding == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> noncoding variants found.</font></i>', sep = '\n')
cat('\n')
```
