
Overview {data-navmenu="SNVs and InDels"}
==================================================================


Row
-------------------------------------

### Overview

```{r}
flexdashboard::valueBox("SNVs/InDels", caption = NULL, color = '#000000', icon = NULL)
```

### Total variants

```{r}
flexdashboard::valueBox(pcg_report$content$snv_indel$vstats$n, color = pcg_report$settings$conf$visual_reporting$color_value_box, icon = "fa-dna")
```


### SNVs

```{r}
flexdashboard::valueBox(pcg_report$content$snv_indel$vstats$n_snv, color = pcg_report$settings$conf$visual_reporting$color_value_box, icon = "fa-dna")
```

### InDels

```{r}
flexdashboard::valueBox(pcg_report$content$snv_indel$vstats$n_indel, color = pcg_report$settings$conf$visual_reporting$color_value_box, icon = "fa-dna")
```

### TIER 1 variants

```{r}
tier1 <- pcg_report$content$snv_indel$vstats$n_tier1
flexdashboard::valueBox(
  tier1, color = ifelse(
    tier1 > 0,
    pcg_report$settings$conf$visual_reporting$color_palette$success$values[1],
    pcg_report$settings$conf$visual_reporting$color_none), icon = "fa-file-prescription", href="#tier-1")
```

### TIER 2 variants

```{r}
tier2 <- pcg_report$content$snv_indel$vstats$n_tier2
flexdashboard::valueBox(
  tier2, color = ifelse(
    tier2 > 0,
    pcg_report$settings$conf$visual_reporting$color_palette$success$values[1],
    pcg_report$settings$conf$visual_reporting$color_none), icon = "fa-file-prescription", href="#tier-2")
```



Row {data-height=300}
-------------------------------------

### Variants per tier {data-width=225}

<br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r pcg_report$settings$conf$visual_reporting[['color_palette']][['tier']][['values']][1]`"><a class="tierlink"  href="#tier-1">&nbsp;TIER 1&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r pcg_report$content$snv_indel$vstats$n_tier1`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r pcg_report$settings$conf$visual_reporting[['color_palette']][['tier']][['values']][2]`"><a class="tierlink"  href="#tier-2">&nbsp;TIER 2&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r pcg_report$content$snv_indel$vstats$n_tier2`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r pcg_report$settings$conf$visual_reporting[['color_palette']][['tier']][['values']][3]`"><a class="tierlink"  href="#tier-3">&nbsp;TIER 3&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r pcg_report$content$snv_indel$vstats$n_tier3`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r pcg_report$settings$conf$visual_reporting[['color_palette']][['tier']][['values']][4]`"><a class="tierlink"  href="#tier-4">&nbsp;TIER 4&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r pcg_report$content$snv_indel$vstats$n_tier4`</font></b><br>

* <font style="color:white; font-size: 125%;font-weight: bold; background-color:`r pcg_report$settings$conf$visual_reporting[['color_palette']][['tier']][['values']][5]`"><a class="tierlink"  href="#tier-5">&nbsp;TIER 5&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r pcg_report$content$snv_indel$vstats$n_tier5`</font></b>


### Allelic support plot {data-padding=10}

```{r vaf_plot_prep1, echo=F}
tumor_af_plot <- F
if(NROW(pcg_report$content$snv_indel$callset$variant) > 0){
  if(NROW(dplyr::filter(
    pcg_report$content$snv_indel$callset$variant, 
    !is.na(DP_TUMOR))) > 0 & 
     NROW(dplyr::filter(
       pcg_report$content$snv_indel$callset$variant, 
       !is.na(AF_TUMOR))) > 0){
    tumor_af_plot <- T
  }
}
```


```{r eval = tumor_af_plot}

af_bin_df <- pcgrr::tier_af_distribution(
  pcg_report$content$snv_indel$callset$variant)
af_bin_df$TIER <- factor(
  af_bin_df$TIER, 
  levels = c("TIER 1","TIER 2","TIER 3","TIER 4","TIER 5"))

p <- ggplot2::ggplot(data = af_bin_df) + 
    ggplot2::geom_bar(
      mapping = ggplot2::aes(
        x = bin_name, y = Count, fill = TIER), 
      stat = "identity") +
    ggplot2::scale_color_manual(values = pcg_report$settings$conf$visual_reporting$color_palette$tier$values) +
    ggplot2::scale_fill_manual(values = pcg_report$settings$conf$visual_reporting$color_palette$tier$values) +
    ggplot2::theme_classic() +
    ggplot2::ylab("Number of variants") +
    ggplot2::xlab("VAF - tumor") + 
  ggplot2::theme(
    legend.title = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(
      angle = 45, 
      family = "Helvetica", 
      size = 9, vjust = -0.1),
    axis.title.x = ggplot2::element_text(
      family = "Helvetica", size = 9, vjust = -2),
    axis.title.y = ggplot2::element_text(
      family = "Helvetica", size = 9, vjust = 1.5))

plotly::ggplotly(p)
```


```{r eval = !tumor_af_plot, results = "asis"}
cat("<br>")
cat('*   <i>Missing variants OR Tumor sequencing depth and allelic fraction of variants not provided - omitting plot.</i>',sep='\n')
cat('<br>')

```

### Documentation {data-padding=20}

The prioritization of SNV and InDels found in the tumor sample is done according to a four-tiered structure, adopting the joint consensus recommendation by AMP/ACMG [Li et al., 2017](https://www.ncbi.nlm.nih.gov/pubmed/27993330).

- [**TIER 1: Variants of strong clinical significance**](#tier-1) - constitutes variants linked to predictive, prognostic, or diagnostic biomarkers in the [CIViC database](http://civic.genome.wustl.edu) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers) that are
	 - Found within the same tumor type/class as specified by the user, **AND**
	 - Of strong clinical evidence (i.e. part of guidelines, validated or discovered in late clinical trials ([CIViC evidence levels A/B](https://civic.readthedocs.io/en/latest/model/evidence/level.html)))
	 - overlap between variants in the tumor sample and reported biomarkers must occur at the _exact variant level_ or at the _codon/exon level_
	 
 - [**TIER 2: Variants of potential clinical significance**](#tier-2) - constitutes other variants linked to predictive, prognostic, or diagnostic biomarkers in the [CIViC database](http://civic.genome.wustl.edu) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers) that are either
	 - Of strong clinical evidence in other tumor types/classes than the one associated with the query sample
	 - Of weak clinical evidence (early trials, case reports etc. ([CIViC evidence levels C/D/E](https://civic.readthedocs.io/en/latest/model/evidence/level.html)))) in the same tumor type/class as the query sample
	 - overlap between variants in the tumor sample and reported biomarkers must occur at the _exact variant level_ or at the _codon/exon level_
	 
 - [**TIER 3: Variants of uncertain clinical significance**](#tier-3) - includes other coding variants found in oncogenes or tumor suppressor genes
   - Status as oncogenes and/or tumor suppressors genes are done according to the following scheme in PCGR:
     - Five or more publications in the biomedical literature that suggests an oncogenic/tumor suppressor role for a given gene (as collected from the [CancerMine text-mining resource](http://bionlp.bcgsc.ca/cancermine/)), **OR**
     - At least two publications from CancerMine that suggests an oncogenic/tumor suppressor role for a given gene **AND** an existing record for the same gene as a tumor suppressor/oncogene in the [Network of Cancer Genes (NCG)](http://ncg.kcl.ac.uk/)
     - Status as oncogene is ignored if a given gene has three times as much (literature evidence) support for a role as a tumor suppressor gene (and vice versa)
     - Oncogenes/tumor suppressor candidates from CancerMine/NCG that are found in the [curated list of false positive cancer drivers compiled by Bailey et al. (Cell, 2018)](https://www.ncbi.nlm.nih.gov/pubmed/30096302) have been excluded
     
 - [**TIER 4**](#tier-4) - includes other protein-coding variants
 - [**TIER 5**](#tier-5) - includes other noncoding variants
 
Row
-----------------------------------------------------

### Global variant datatable - filters {data-padding=10 data-width=450}

The global variant datatable (right) can be filtered according to various criteria:
<br>

* Filtering on sequencing depth/variant allelic fraction depends on input provided by user
* Filtering performed here will only apply to the datatable and not any other visualizations presented in this page

<br>

```{r table_browse_allelic_depth, echo=F, results = "asis", eval = NROW(pcg_report[['content']][['snv_indel']][['variant_set']][['tsv']]) > 0}

if(NROW(pcg_report[['content']][['snv_indel']][['callset']][['variant']]) > 2000){
  cat('<b>NOTE - listing top 2000 variants</b><br>',sep="\n")
  cat('<br>')
  pcg_report[['content']][['snv_indel']][['callset']][['variant']] <- head(pcg_report[['content']][['snv_indel']][['callset']][['variant']], 2000)
}


vars <- pcg_report[['content']][['snv_indel']][['callset']][['variant']] |> 
  dplyr::select(SYMBOL, CONSEQUENCE, PROTEIN_CHANGE, TIER, VARIANT_CLASS,
                GENOMIC_CHANGE, GENOME_VERSION, DP_TUMOR, 
                AF_TUMOR, CALL_CONFIDENCE, DP_CONTROL, AF_CONTROL) |>
  dplyr::mutate(CONSEQUENCE = stringr::str_replace_all(CONSEQUENCE,"&",","))

tiers <- c('TIER 1','TIER 2','TIER 3','TIER 4','TIER 5')

vars <- pcg_report[['content']][['snv_indel']][['callset']][['variant']] |> 
  dplyr::select(
    dplyr::any_of(
      c("SYMBOL", "CONSEQUENCE", "PROTEIN_CHANGE", 
        "VARIANT_CLASS", "TIER", "GENOMIC_CHANGE", 
        "GENOME_VERSION", "CALL_CONFIDENCE", "DP_TUMOR", 
        "AF_TUMOR", "DP_CONTROL", "AF_CONTROL")
      )) |>
  dplyr::mutate(
    CONSEQUENCE = stringr::str_replace_all(CONSEQUENCE,"&",",")) |>
  dplyr::mutate(
    TIER = paste0("TIER ", TIER)
  )

variants_shared <- crosstalk::SharedData$new(vars)


filter_set1 <- list()
filter_set2 <- list()

filter_set1[[1]] <- crosstalk::filter_select(
  "TIER", "Tier", variants_shared, ~TIER)
filter_set2[[1]] <- crosstalk::filter_select(
  "CONSEQUENCE", "Consequence", variants_shared, ~CONSEQUENCE)

i_set1 <- 2
i_set2 <- 2
if(any(!is.na(vars$CALL_CONFIDENCE))){
  filter_set1[[i_set1]] <- crosstalk::filter_select(
    "CALL_CONFIDENCE", "Call confidence", variants_shared, ~CALL_CONFIDENCE)
  i_set1 <- i_set1 + 1
}
if(any(!is.na(vars$DP_TUMOR)) & any(!is.na(vars$AF_TUMOR))){
  filter_set1[[i_set1]] <- crosstalk::filter_slider(
    "DP_TUMOR", "Sequencing depth tumor", 
    variants_shared, ~DP_TUMOR, ticks = F, sep = F)
  filter_set2[[i_set2]] <- crosstalk::filter_slider(
    "AF_TUMOR", "Allelic fraction tumor", 
    variants_shared, ~AF_TUMOR, 
    sep = F, min = 0, max = 1, step = 0.05, ticks = T)
  i_set1 <- i_set1 + 1
  i_set2 <- i_set2 + 1
}
if(any(!is.na(vars$DP_CONTROL)) & any(!is.na(vars$AF_CONTROL))){
  filter_set1[[i_set1]] <- crosstalk::filter_slider(
    "DP_CONTROL", "Sequencing depth control", 
    variants_shared, ~DP_CONTROL, ticks = F, sep = F)
  filter_set2[[i_set2]] <- crosstalk::filter_slider(
    "AF_CONTROL", "Allelic fraction control", 
    variants_shared, ~AF_CONTROL, 
    sep = F, min = 0, max = 1, step = 0.05, ticks = T)
}

crosstalk::bscols(
  filter_set1,
  filter_set2
)

```


### Global variant datatable {data-padding=15}

```{r table_browse, eval = NROW(pcg_report[['content']][['snv_indel']][['variant_set']][['tsv']]) > 0}

tiers <- c('TIER 1','TIER 2','TIER 3','TIER 4','TIER 5')

variants_shared |>
  DT::datatable(
    extensions=c("Buttons","Responsive"),
    #plugins = "scrollResize",
    options = list(buttons = c('csv','excel'),  
                 pageLength = 8, 
                 ##autoWidth = T, 
                 #width = "100%",
                 #fixedColumns = T,
                 #fixedHeader = T,
                 #autoWidth = TRUE,
                  #columnDefs = list(list(width = '80px', targets = c(1,3,4)))
                 scrollX = "900px",
                 scrollCollapse = T,
                 scrollY = scrollY_flexdb_container,
                  #columnDefs = list(list(width = '6%', targets = c(1,3,4,5)),
                  #                  list(width = '12%', targets = c(2,6))),
                  # columnDefs = list(list(width = '80px', targets = c(1,3,4)),
                  #                   list(width = '90px', targets = 6),
                  #                   list(width = '40px', targets = 2),
                  #                   list(width = '40px', targets = 5)),
                 dom = 'Bfrtip')
  ) |>
   DT::formatStyle('TIER',color = 'white', 
                   backgroundColor = DT::styleEqual(tiers, head(pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']],5)), 
                   fontWeight = 'bold', `text-align` = 'center')


```
