TMB & MSI
================================================================== 

```{r check_analysis_status_msi, echo = F}
missing_data_norun <- F
if(pcg_report$content$msi$eval == F | pcg_report$content$msi$missing_data == T){
  missing_data_norun <- T
}
```

Row
-------------------------------------

### Tumor mutational burden & MSI classification

```{r width = 5, tmb_vbox_overview}
flexdashboard::valueBox("TMB/MSI status", caption = NULL, color = '#000000', icon = NULL)
```

### Coding target size

```{r tmb_vbox_size, width = 4, echo = F}
col <- pcg_report$settings$conf$visual_reporting$color_value_box
if(pcg_report$metadata$config$assay_props$target_size_mb < 1){
  col <- pcg_report$settings$conf$visual_reporting$color_palette[['warning']][['values']][1]
}

flexdashboard::valueBox(paste0(pcg_report$metadata$config$assay_props$target_size_mb," Mb"),
                        color = col,
                        icon = "fa-dna")
```

### Coding mutations

```{r tmb_n_muts, width = 4, echo = F}
flexdashboard::valueBox(pcg_report$content$snv_indel$vstats$n_coding,
                        color = pcg_report$settings$conf$visual_reporting$color_value_box,
                        icon = "fa-dna")
```


### TMB estimate

```{r tmb_vbox_estimate, echo = F}
flexdashboard::valueBox(paste0(pcg_report$content$tmb$vstats$tmb_estimate," mut/Mb"), 
                        color = pcg_report$settings$conf$visual_reporting$color_value_box,
                        icon = "fa-dna")
```

### MSI prediction

```{r msi_classification, echo = F}

msi_status <- 'ND'
if(!missing_data_norun){
  msi_status <- stringr::str_replace(pcg_report$content$msi$prediction$msi_stats$vb,"MSI status:\n","")
}

flexdashboard::valueBox(msi_status, 
                        color = pcg_report$settings$conf$visual_reporting$color_value_box,
                        icon = "fa-dna")
```


Row {data-height=500}
-------------------------------------

### TCGA TMB distribution {data-width=500 data-padding=15}

```{r tmb_plot, echo = F}
tcga_tmb <- pcg_report[['content']][['tmb']][['tcga_tmb']]
p_site <- pcg_report[['metadata']][['config']][["t_props"]][['tumor_type']]
tmb_estimate = pcg_report[['content']][['tmb']][['v_stat']][['tmb_estimate']]
algorithm <- pcg_report[['content']][['tmb']][['algorithm']]
#tmb_high = pcg_report[['metadata']][['config']][['tmb']][['tmb_intermediate_limit']]

tmb_site_colors <- 
  data.frame(primary_site = unique(tcga_tmb$primary_site), stringsAsFactors = F) |>
    dplyr::filter(!is.na(primary_site)) |>
  dplyr::mutate(color = "#f0f0f0") |>
  dplyr::mutate(color = dplyr::if_else(primary_site == p_site,
                                       pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][1], color))
  
tmb_site_color_vec <- tmb_site_colors$color
names(tmb_site_color_vec) <- tmb_site_colors$primary_site

tcga_tmb <- tcga_tmb |> dplyr::filter(!is.na(primary_site))

if(algorithm == "nonsyn"){
  tcga_tmb <- tcga_tmb |> dplyr::filter(!is.na(primary_site)) |>
    dplyr::select(primary_site, tmb_ns_log10, tmb_ns) |>
    dplyr::rename(tmb = tmb_ns, tmb_log10 = tmb_ns_log10)
}


tmb_plot_site <-
  #ggplot2::ggplot(data = tcga_tmb[['tmb_site']]) +
  ggplot2::ggplot(data = tcga_tmb) +
  ggplot2::geom_boxplot(mapping = ggplot2::aes(x = reorder(primary_site, tmb_log10, FUN=median), 
                                               y = tmb, fill = primary_site), outlier.size = 0.2) +
  ggplot2::theme_classic() +
  ggplot2::scale_y_continuous(trans = scales::log_trans(base=10), 
                              breaks=c(0.01,1,10,100,1000),
                              labels = c("0.01","1","10","100","1000")) +
  ggplot2::scale_fill_manual(values = tmb_site_color_vec) +
  ggplot2::xlab("Primary Site") +
  ggplot2::ylab("TMB (mutations/mb)") +
  ggplot2::theme(legend.title = ggplot2::element_blank(),
                 axis.title.y = ggplot2::element_blank(),
                 axis.title.x = ggplot2::element_blank(),
                 legend.position = "none",
                 axis.text.x = ggplot2::element_text(family = "Helvetica", angle = 55, vjust=0.5, size = 10),
                 axis.text.y = ggplot2::element_text(family = "Helvetica", size = 10),
                 plot.margin = (grid::unit(c(0, 0, 0, 0), "cm"))) + 
                 ggplot2::geom_hline(yintercept=as.numeric(tmb_estimate), size=0.9, 
                                     linetype = 4, color = pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][1])
                  #ggplot2::geom_rect(ggplot2::aes(ymin = tmb_high, ymax = 1020, xmin = -Inf, xmax = Inf),
                     #fill = "gray", alpha = 0.01)

plotly::ggplotly(tmb_plot_site)
```

### MSI evidence I {data-width=500 data-padding=15}

<br>


```{r msi_desc1, echo = F, eval = pcg_report$content$msi$eval, results = 'asis'}

cat(paste0('The plot below illustrates the <i><b>mutational burden of indels</b></i> in ',pcg_report$metadata$sample_name,' (black dashed line) along with the distribution in TCGA samples for samples with known MSI status ( <i>MSI.H</i> = high microsatellite instability, <i>MSS</i> = microsatellite stable):'))
```

```{r msi_desc2, echo = F, eval = missing_data_norun, results = 'asis'}
cat('\n*  <i><font style="font-size: 110%"><b>NOT DETERMINED (ND)/SWITCHED OFF</b> - evaluated for tumor-control WGS/WES assays only</i></font>', sep = '\n')
cat('\n')
```
    
  
```{r msi_indel_tmb_plot, echo = F, fig.height = 2.5, fig.width = 5, eval = pcg_report$content$msi$eval}
indel_tmb <- 
  pcg_report[['content']][['msi']][['prediction']][['msi_stats']][['tmb_indel']]

p <- ggplot2::ggplot(data = pcg_report[['content']][['msi']][['prediction']][['tcga_dataset']]) +
    ggplot2::geom_boxplot(mapping = ggplot2::aes(y = tmb_indel, x = MSI_status, color = MSI_status, fill = MSI_status),
                            position = "dodge") +
    #ggplot2::ylab("Number of TCGA samples") +
    ggplot2::theme_classic() +
    ggplot2::scale_color_manual(values = c(pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][2], pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][1])) +
    ggplot2::scale_fill_manual(values = c(pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][2], pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][1])) +
    ggplot2::ylab("Burden of indels (indels/Mb)") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(family = "Helvetica", size = 11),
                   axis.title.x = ggplot2::element_blank(),
                   legend.title = ggplot2::element_blank(),
                   legend.text = ggplot2::element_text(family = "Helvetica", size = 11),
                   axis.text.y = ggplot2::element_text(family = "Helvetica", size = 11),
                   axis.title.y = ggplot2::element_text(family = "Helvetica", size = 11, vjust = 1.5),
                   plot.margin = ggplot2::margin(.5, .5, .5, 1.5, "cm")) +
    ggplot2::geom_hline(yintercept = as.numeric(indel_tmb), size = 0.9, linetype = "dashed")

p


htmltools::br()
```

Row
-------------------------------------

### MSI evidence II {data-width=500 data-padding=15}

<br>


```{r msi_desc3, echo = F, eval = pcg_report$content$msi$eval, results = 'asis'}

cat(paste0('The plot below illustrates the <i><b>fraction of indels among all calls</b></i> in ',pcg_report$metadata$sample_name, '(black dashed line) along with the distribution in TCGA for samples with known MSI status ( <i>MSI.H</i> = high microsatellite instability, <i>MSS</i> = microsatellite stable):'))

```

```{r msi_desc4, echo = F, eval = missing_data_norun, results = 'asis'}
cat('\n*  <i><font style="font-size: 110%"><b>NOT DETERMINED (ND)/SWITCHED OFF</b> - evaluated for tumor-control WGS/WES assays only</i></font>', sep = '\n')
cat('\n')
```

```{r msi_indelfrac_plot, echo = F, fig.height = 2.5, fig.width = 6, eval = pcg_report$content$msi$eval}

indel_fraction <- 
  pcg_report[['content']][['msi']][['prediction']][['msi_stats']]$fracIndels

p <- ggplot2::ggplot(data = pcg_report[['content']][['msi']][['prediction']][['tcga_dataset']]) +
    ggplot2::geom_histogram(mapping = ggplot2::aes(x = fracIndels, color = MSI_status, fill = MSI_status),
                            position = "dodge", binwidth = 0.01) +
    ggplot2::ylab("Number of TCGA samples") +
    ggplot2::theme_classic()+ 
    ggplot2::scale_color_manual(values = c(pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][2], pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][1])) +
    ggplot2::scale_fill_manual(values = c(pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][2], pcg_report$settings$conf$visual_reporting$color_palette[['tier']][['values']][1])) +
    ggplot2::xlab("InDel fraction among somatic calls") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(family = "Helvetica", size = 11),
                   axis.title.x = ggplot2::element_text(family = "Helvetica", size = 11),
                   legend.title = ggplot2::element_blank(),
                   legend.text = ggplot2::element_text(family = "Helvetica", size = 11),
                   axis.text.y = ggplot2::element_text(family = "Helvetica", size = 11),
                   axis.title.y = ggplot2::element_text(family = "Helvetica", size = 11, vjust = 1.5),
                   plot.margin = ggplot2::margin(0.5, 0.5, 0.5, 0.5, "cm")) +
    ggplot2::geom_vline(xintercept = as.numeric(indel_fraction), size = 0.9, linetype = "dashed")

p

htmltools::br()
```

### Somatic mutations in MMR genes {data-width=500 data-padding=20}

<br>


```{r msi_desc5, echo = F, eval = missing_data_norun, results = 'asis'}
cat('\n*  <i><font style="font-size: 110%"><b>NOT DETERMINED (ND)/SWITCHED OFF</b> - evaluated for tumor-control WGS/WES assays only</i></font>', sep = '\n')
cat('\n')
```


```{r msi_gene_variants, echo=F, results='asis', eval = pcg_report$content$msi$eval}
if(NROW(pcg_report[['content']][['msi']][['prediction']][['mmr_pol_variants']]) > 0){
  DT::datatable(pcg_report[['content']][['msi']][['prediction']][['mmr_pol_variants']], 
                escape=F,extensions=c("Buttons","Responsive"), 
                width = "100%",
                options=list(
                  buttons = c('csv','excel'), 
                  scrollY = scrollY_flexdb_container, 
                  #autoWidth = T, 
                  #fixedColumns = T,
                  #fixedHeader = T,
                  scrollX = "900px",
                  #columnDefs = list(list(width = '90px', targets = c(1,2,3,4,5))),
                  dom = 'Bfrtip'))
}else{
  cat('<i>No variants found.</i>',sep='\n')
  cat('\n')
}
```
<br><br>
