Tier 1 {data-navmenu="SNVs and InDels"}
==================================================================

Row
-------------------------------------

### SNVs and InDels

```{r}
flexdashboard::valueBox(
  "TIER 1", caption = NULL, color = '#000000', icon = NULL)
```

### Biomarker genes

```{r}

flexdashboard::valueBox(
  pcg_report$content$snv_indel$vstats$n_genes_tier1, 
  color = pcg_report$settings$conf$report_color)

```

### Biomarker variants

```{r}
flexdashboard::valueBox(
  pcg_report$content$snv_indel$vstats$n_tier1, 
  color =  pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```


### Diagnostic evidence items

```{r}
flexdashboard::valueBox(
  pcg_report$content$snv_indel$vstats$n_eitems_diagnostic_tier1, 
  color = ifelse(
    pcg_report$content$snv_indel$vstats$n_eitems_diagnostic_tier1 > 0,
    pcgrr::color_palette$success,
    pcgrr::color_palette$none), 
  icon = "fa-file-prescription")
```

### Prognostic evidence items

```{r}
flexdashboard::valueBox(
  pcg_report$content$snv_indel$vstats$n_eitems_prognostic_tier1, 
  color = ifelse(
    pcg_report$content$snv_indel$vstats$n_eitems_prognostic_tier1 > 0,
    pcgrr::color_palette$success,
    pcgrr::color_palette$none), 
  icon = "fa-file-prescription")
```

### Predictive evidence items

```{r}
flexdashboard::valueBox(
  pcg_report$content$snv_indel$vstats$n_eitems_predictive_tier1, 
  color = ifelse(
    pcg_report$content$snv_indel$vstats$n_eitems_predictive_tier1 > 0,
    pcgrr::color_palette$success,
    pcgrr::color_palette$none), 
  icon = "fa-file-prescription")
```


Row
-----------------------------------------------------

### Tier 1 variant evidence items - filters {data-padding=15 data-width=450}

<br>

```{r table_browse_tier1, echo=F, results = "asis", eval = pcg_report$content$snv_indel$vstats$n_actionability_tier1 > 0}

cat('\nEvidence items associated with variants in tier 1 (right panel) can be filtered according to various criteria:\n')

var_eitems_tier1 <- pcg_report$content$snv_indel$callset$variant_display |>
  dplyr::filter(!is.na(.data$TIER) & TIER == 1) |>
  dplyr::select(-c("TIER_DESCRIPTION")) |>
  dplyr::inner_join(
    pcg_report$content$snv_indel$callset$biomarker_evidence$items,
    by = c("VAR_ID","TIER","VARIANT_CLASS","ENTREZGENE")
  ) |>
  dplyr::filter(.data$BM_RESOLUTION != "gene") |>
  dplyr::distinct() |>
  dplyr::select(
    dplyr::any_of(
      pcgrr::dt_display$var_eitem)) |>
  dplyr::distinct()

var_eitems_tier1_shared <- crosstalk::SharedData$new(
  var_eitems_tier1)

crosstalk::bscols(
  list(
    crosstalk::filter_select(
      "BM_CANCER_TYPE", 
      "Cancer type", 
      var_eitems_tier1_shared, 
      ~BM_CANCER_TYPE),
    crosstalk::filter_select(
      "BM_CLINICAL_SIGNIFICANCE", 
      "Clinical significance", 
      var_eitems_tier1_shared, 
      ~BM_CLINICAL_SIGNIFICANCE),
    crosstalk::filter_select(
      "BM_EVIDENCE_LEVEL", 
      "Evidence level", 
      var_eitems_tier1_shared, 
      ~BM_EVIDENCE_LEVEL),
    crosstalk::filter_select(
      "BM_EVIDENCE_DIRECTION", 
      "Evidence direction", 
      var_eitems_tier1_shared, 
      ~BM_EVIDENCE_DIRECTION),
    crosstalk::filter_slider(
      "BM_RATING", 
      "Rating", 
      var_eitems_tier1_shared, 
      ~BM_RATING, min = 0, max = 5, 
      step = 1, ticks = T)

  ),
  list(
    crosstalk::filter_select(
      "CONSEQUENCE", 
      "Consequence", 
      var_eitems_tier1_shared, 
      ~CONSEQUENCE),
    crosstalk::filter_select(
      "BM_EVIDENCE_TYPE", 
      "Evidence type", 
      var_eitems_tier1_shared, 
      ~BM_EVIDENCE_TYPE),
    crosstalk::filter_select(
      "BM_RESOLUTION", 
      "Biomarker resolution", 
      var_eitems_tier1_shared, 
      ~BM_RESOLUTION),
    crosstalk::filter_select(
      "BM_THERAPEUTIC_CONTEXT", 
      "Therapeutic context", 
      var_eitems_tier1_shared, 
      ~BM_THERAPEUTIC_CONTEXT)
  )
)


```


```{r biomarker_note1, echo=F, results = "asis", include = pcg_report$content$snv_indel$vstats$n_actionability_tier1 > 0}

cat('<b>NOTE:</b> Reported biomarkers in CIViC/CGI are mapped at different resolutions (i.e. filter <b>Biomarker resolution</b>). The accuracy of a match between variants in the tumor sample and the reported biomarkers will vary accordingly (highlighted by gene symbols with different color backgrounds):\n\n')

cat('<ul><li>Biomarker match at the <mark style="background-color:black; font-weight:bold; color:white">exact genomic, amino acid or codon level</mark></li>')
cat(paste0('<br><li>Biomarker match at the <mark style="background-color:', pcgrr::color_palette$warning,'; color:white; font-weight:bold">exon/gene level</mark></li></ul>\n'))

htmltools::br()
htmltools::br()
```


```{r tier1_missing_filters, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$vstats$n_actionability_tier1 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> variants of strong clinical significance found.</i></font>', sep='\n')
cat('\n')
```


### Tier 1  - variant evidence items {data-padding=15}

<br>

```{r table_tier1, eval = pcg_report$content$snv_indel$vstats$n_actionability_tier1 > 0}

warning_color <- 
  pcgrr::color_palette$warning

var_eitems_tier1_shared |>
  DT::datatable(
    escape = F, 
    extensions = c("Buttons","Responsive"), 
    options = list(
      pageLength = 8, 
      scrollY = scrollY_flexdb_container, 
      buttons = c('csv','excel'), 
      fixedColumns = T,
      fixedHeader = T,
      width = "100%",
      scrollX = "900px",
      columnDefs = list(
        list(width = '90px', 
             targets = c(1,2,3,4,5,6))),
      dom = 'Bfrtip')) |>
  DT::formatStyle(
    'BM_EVIDENCE_LEVEL', 
    backgroundColor = DT::styleEqual(
      pcgrr::color_palette$clinical_evidence$levels,
      pcgrr::color_palette$clinical_evidence$values)) |>
  DT::formatStyle(
    color="white", 
    "SYMBOL", 
    "BM_RESOLUTION", 
    fontWeight = 'bold', 
    `text-align` = 'center',
    backgroundColor = DT::styleEqual(
      c('genomic',
        'hgvsp',
        'codon',
        'exon',
        'gene'), 
      c('#000',
        '#000',
        '#000',
        warning_color, 
        warning_color)))

```


```{r tier1_missing_data, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$vstats$n_actionability_tier1 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> variants of strong clinical significance found.</i></font>',sep='\n')
cat('\n')
```

