---
title: "Cancer Genome Sequencing Report by PCGR"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: bibliography.bib
nocite: |
  @Tamborero2018-aj, @Poti2019-qr, @Alexandrov2020-gj, @Alexandrov2013-gs, @Griffith2017-do, @Ma2018-to, @Cortes-Ciriano2017-vl, @Li2017-ew,
  @Blokzijl2018-nc, @Lever2019-xp, @Repana2019-dd, @Chalmers2017-cz, @Yin2020-or, @Bailey2018-os, @Fernandez2019-mk
output: 
  html_document:
    includes:
      in_header: header.html
      after_body: disclaimer.md
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = F,warning=FALSE, dpi=72, error=F, eval=T)
```

```{r settings_data_sources, child='pcgr_rmarkdown/report_query_settings.Rmd'}
```

```{r conditional_value_box, child='pcgr_rmarkdown/value_boxes.Rmd', eval = pcg_report[['content']][['value_box']][['eval']] & pcg_report[['metadata']][['config']][['assay_props']][['vcf_tumor_only']] == F}
```


```{r conditional_value_box, child='pcgr_rmarkdown/value_boxes_tumor_only.Rmd', eval = pcg_report[['content']][['value_box']][['eval']] & pcg_report[['metadata']][['config']][['assay_props']][['vcf_tumor_only']] == T}
```

<!--
```{r conditional_seq_mode, child='pcgr_rmarkdown/snv_indel_header.Rmd', eval = pcg_report[['content']][['sequencing_mode']][['eval']]}
```
-->

## Somatic SNVs/InDels

```{r no_variants, eval = pcg_report[['content']][['snv_indel']][['zero']], results = "asis"}
htmltools::br()>

cat('\nThere are __0__ _PASS_ variants (potentially after filtering) in the input VCF file\n',sep="\n")

htmltools::br()
```

```{r conditional_tumor_only, child='pcgr_rmarkdown/tumor_only_statistics.Rmd', eval = pcg_report[['content']][['tumor_only']][['eval']]}
```

```{r conditional_mutational_burden, child='pcgr_rmarkdown/mutational_burden.Rmd', eval = pcg_report[['content']][['tmb']][['eval']]}
```

```{r conditional_var_stats, child='pcgr_rmarkdown/global_table_browse.Rmd', eval = pcg_report[['content']][['snv_indel']][['eval']]}
```

```{r conditional_tier1, child='pcgr_rmarkdown/snv_tier1.Rmd', eval = pcg_report[['content']][['snv_indel']][['eval']]}
```

```{r conditional_tier2, child='pcgr_rmarkdown/snv_tier2.Rmd', eval = pcg_report[['content']][['snv_indel']][['eval']]}
```

```{r conditional_tier3, child='pcgr_rmarkdown/snv_tier3.Rmd', eval = pcg_report[['content']][['snv_indel']][['eval']]}
```

```{r conditional_tier4, child='pcgr_rmarkdown/snv_tier4.Rmd', eval = pcg_report[['content']][['snv_indel']][['eval']]}
```

```{r conditional_tier_noncoding, child='pcgr_rmarkdown/snv_noncoding.Rmd', eval = pcg_report[['metadata']][['config']][['other']][['list_noncoding']]}
```

```{r conditional_tier_noncoding, child='pcgr_rmarkdown/snv_biomarker.Rmd', eval = pcg_report[['content']][['snv_indel']][['eval']]}
```


```{r conditional_cna_overview, child='pcgr_rmarkdown/cna_segments.Rmd', eval = pcg_report[['content']][['cna']][['eval']]}
```

```{r conditional_cna_oncogene, child='pcgr_rmarkdown/cna_oncogene_gain.Rmd', eval = pcg_report[['content']][['cna']][['eval']]}
```

```{r conditional_cna_tsgene, child='pcgr_rmarkdown/cna_tsgene_loss.Rmd', eval = pcg_report[['content']][['cna']][['eval']]}
```

```{r conditional_cna_drug_targets, child='pcgr_rmarkdown/cna_targets_gain.Rmd', eval = pcg_report[['content']][['cna']][['eval']]}
```

```{r conditional_cna_biomarker, child='pcgr_rmarkdown/cna_biomarkers.Rmd', eval = pcg_report[['content']][['cna']][['eval']]}
```

```{r conditional_msi, child='pcgr_rmarkdown/msi.Rmd', eval = pcg_report[['content']][['msi']][['eval']]}
```

```{r conditional_msi_missing, eval = pcg_report[['content']][['msi']][['missing_data']], results = "asis"}
cat("\n")

cat("## MSI status")

cat("\nThe number of SNV/InDel calls (n = ",pcg_report[['content']][['snv_indel']][['v_stat']][['n_coding']], 
"is too limited for MSI prediction to be performed.",sep="\n")

htmltools::br()
htmltools::br()

```

```{r conditional_msig, child='pcgr_rmarkdown/mutational_signature_mp.Rmd', eval = pcg_report$metadata$config$msigs$run == T &  !pcg_report[['content']][['mutational_signatures']][['missing_data']]}
```

```{r conditional_msig_missing, eval = pcg_report$metadata$config$msigs$run == T & pcg_report[['content']][['mutational_signatures']][['missing_data']], results = "asis"}

cat("\n")

cat("## Mutational signatures")

cat("\nThe number of SNV calls (n = ",nrow(pcg_report[['content']][['mutational_signatures']][['variant_set']][['all']]),") is too limited. A minimum of ",pcg_report[['metadata']][['config']][['msigs']][['mutation_limit']]," SNVs is currently required for a mutational signature analysis to be performed.")

htmltools::br()
htmltools::br()
```

```{r conditional_kataegis, child='pcgr_rmarkdown/kataegis.Rmd', eval = pcg_report[['content']][['kataegis']][['eval']]}
```
<br><br>


```{r conditional_cpsr, child='pcgr_rmarkdown/cpsr_main_findings.Rmd', eval = pcg_report[['content']][['cpsr']][['eval']]}
```


```{r conditional_trials, child='pcgr_rmarkdown/clinicaltrials.Rmd', eval = pcg_report[['content']][['clinicaltrials']][['eval']]}
```


```{r documentation, child='pcgr_rmarkdown/documentation.Rmd', eval = TRUE}
```


### References

