Tier 3 {data-navmenu="SNVs and InDels"}
================================================================== 

Row
-------------------------------------

### SNVs and InDels

```{r}
flexdashboard::valueBox("TIER 3", caption = NULL, color = '#000000', icon = NULL)
```

### Total variants

```{r}
flexdashboard::valueBox(
  pcg_report$content$snv_indel$vstats$n_actionability_tier3, 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```

### Variants in proto-oncogenes

```{r oncogene_variants, echo = F}
flexdashboard::valueBox(
  as.character(
    pcg_report$content$snv_indel$vstats$n_actionability_tier3_oncogene), 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```


### Variants in tumor suppressor genes

```{r tsgene_variants, echo = F}
flexdashboard::valueBox(
  as.character(
    pcg_report$content$snv_indel$vstats$n_actionability_tier3_tsg), 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```

### Variants in genes with dual roles

```{r dual_role_variants, echo = F}
flexdashboard::valueBox(
  as.character(
    pcg_report$content$snv_indel$vstats$n_actionability_tier3_dualrole), 
  color = pcg_report$settings$conf$report_color, 
  icon = "fa-dna")
```


Row
-----------------------------------------------------

### Tier 3 - variant filters {data-padding=15 data-width=250}

<br>

```{r table_browse_tier3, echo=F, results = "asis", eval = pcg_report$content$snv_indel$vstats$n_tier3 > 0}

tier3_vars <- pcg_report$content$snv_indel$callset$variant_display |>
  dplyr::filter(!is.na(.data$TIER) & 
                  .data$TIER == 3) |>
  dplyr::select(-c("TIER_DESCRIPTION")) |>
  dplyr::distinct() |>
  dplyr::select(
    dplyr::any_of(
      pcgrr::dt_display$tier3))

oncogenicity_levels <- levels(as.factor(tier3_vars$ONCOGENICITY))

cat('Variants in TIER 3 (right panel) can be interactively explored according to various criteria\n:')

variants_tier3_shared <- crosstalk::SharedData$new(tier3_vars)
crosstalk::bscols(
  list(
    crosstalk::filter_select(
      "CONSEQUENCE", "Consequence", 
      variants_tier3_shared, ~CONSEQUENCE),
    crosstalk::filter_select(
      "TUMOR_SUPPRESSOR", "Tumor suppressor", 
      variants_tier3_shared, ~TUMOR_SUPPRESSOR),
    crosstalk::filter_select(
      "ONCOGENE", "Proto-oncogene", 
      variants_tier3_shared, ~ONCOGENE)
  ),
  list(
    crosstalk::filter_select(
      "SYMBOL", "Gene", 
      variants_tier3_shared, ~SYMBOL),
    crosstalk::filter_select(
      "MUTATION_HOTSPOT", "Mutation hotspot", 
      variants_tier3_shared, ~MUTATION_HOTSPOT),
    crosstalk::filter_select(
      "ONCOGENICITY", "Oncogenicity", 
      variants_tier3_shared, ~ONCOGENICITY)

    
  )
)

```


```{r tier3_missing_filters, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$vstats$n_tier3 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> other variants found in proto-oncogenes or tumor suppressor genes.</font></i>', sep = '\n')
cat('\n')
```


### Tier 3  - variant datatable {data-padding=15}

<br>


```{r table_tier3, eval = pcg_report$content$snv_indel$vstats$n_actionability_tier3 > 0}

dt <- DT::datatable(
  variants_tier3_shared, 
  escape = F,
  extensions = c("Buttons","Responsive"), 
  options = 
    list(pageLength = 8, 
         scrollY = scrollY_flexdb_container, 
         buttons = c('csv','excel'), 
         width = "100%",
         scrollX = F,
         dom = 'Bfrtip')) |>
  DT::formatStyle(
    "SYMBOL",
    "GLOBAL_ASSOC_RANK", 
    color = "white", 
    backgroundColor = 
      DT::styleInterval(
        pcgrr::color_palette$cancer_assoc$breaks, 
        pcgrr::color_palette$cancer_assoc$values))

if(length(oncogenicity_levels) > 0){
  dt <- dt |>
    DT::formatStyle(
      "ONCOGENICITY",
      "ONCOGENICITY",
      color = "white", 
      fontWeight = "bold",
      backgroundColor = DT::styleEqual(
        pcgrr::color_palette$oncogenicity$levels,
        pcgrr::color_palette$oncogenicity$values)
    )
}
dt

```

```{r tier3_missing_data, echo=F, results = 'asis', eval = pcg_report$content$snv_indel$vstats$n_actionability_tier3 == 0}
cat('\n*  <i> <font style="font-size: 110%"><b>No</b> other variants found in proto-oncogenes or tumor suppressor genes.</font></i>', sep = '\n')
cat('\n')
```

