## Documentation

This report is intended for interpretation of acquired genomic aberrations  detected in human tumor samples, as detected from high-throughput sequencing
technologies.

<br><br>

::: {.panel-tabset}

### Gene/variant annotation resources {#sec-resources}

The variant interpretation procedures performed in this cancer genome report are relying upon multiple tools and knowledge resources, as outlined below. 

::: {.callout-important}
## Licensing note

Some resources (highlighted below) require specific licensing agreements if you plan to utilize their data (and thus this report) in a commercial, non-research setting.
:::


* __PCGR databundle version__

  * __`r pcg_report$settings$reference_data$version`__

* __Databases/datasets__ 
```{r list_db}
#| echo: false
#| output: asis

i <- 1
ref_datasets <- pcg_report$settings$reference_data$source_metadata
for(i in 1:NROW(ref_datasets)){
  source_full <- ref_datasets[i,"source"]
  source <- ref_datasets[i,"source_abbreviation"]
  description <- ref_datasets[i,"source_description"]
  url <- ref_datasets[i,"source_url"]
  version <- ref_datasets[i,"source_version"]
  license <- ref_datasets[i, "source_license"]
  license_url <- ref_datasets[i, "source_license_url"]
  
  ## temporary fix for outdated ACMG URL (geneOncoX needs update):
  if(source == "acmg_sf"){
    url <- "https://pubmed.ncbi.nlm.nih.gov/37347242/"
  }
  wflow <- ref_datasets[i, "wflow"]
  if(!(stringr::str_detect(
    wflow,"pcgr"))){
    next
  }
  
  s <- ""
  if(is.na(version)){
    if(!is.na(license_url)){
      s <- paste0("    * [", source_full, "](", url, ") - ", description, " - [",
                  license,"](", license_url, ")")
    }else{
      s <- paste0("    * [", source_full, "](", url, ") - ", description, " - ",
                  license)
    }
  }else{
    if(!is.na(license_url)){
      if(source == "cgc" | source == "gepa"){
        s <- paste0("    * [", source_full, "](", url, ") - ", description, " (<b>", version, "</b>)",
                    " - <b>[", license,"](", license_url, ")</b>")
      }else{
        s <- paste0("    * [", source_full, "](", url, ") - ", description, " (<b>", version, "</b>)",
                    " - [", license,"](", license_url, ")")
      }
    }else{
       s <- paste0("    * [", source_full, "](", url, ") - ", description, " (<b>", version, "</b>)",
                  " - ", license)
    }
  }
  if(s != ""){
    cat(s, sep = "\n")
  }
}

```

### Report content

#### Actionability classificaton

Clinical actionability assessment of SNVs/InDels and gene copy number aberrations found in the tumor sample follows recommendation guidelines by ACMG/AMP [@Li2017-ew]. Specifically, different levels  of actionability are implemented in the following manner:

- **TIER 1: Variants of strong clinical significance** - constitutes aberrations linked to predictive, prognostic, or diagnostic biomarkers in the [CIViC database](http://civic.genome.wustl.edu) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers) that are
	 - Found within the same tumor type/class as specified by the user, **AND**
	 - Of strong clinical evidence (i.e. part of guidelines, validated or discovered in late clinical trials ([CIViC evidence levels A/B](https://civic.readthedocs.io/en/latest/model/evidence/level.html)))
	 - overlap between variants in the tumor sample and reported biomarkers must occur at the _exact variant level_ or at the _codon/exon level_
	 
 - **TIER 2: Variants of potential clinical significance** - constitutes other aberrations linked to predictive, prognostic, or diagnostic biomarkers in the [CIViC database](http://civic.genome.wustl.edu) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers) that are either
	 - Of strong clinical evidence in other tumor types/classes than the one specified by the user, **OR**
	 - Of weak clinical evidence (early trials, case reports etc. ([CIViC evidence levels C/D/E](https://civic.readthedocs.io/en/latest/model/evidence/level.html)))) in the same tumor type/class as specified by the user
	 - overlap between variants in the tumor sample and reported biomarkers must occur at the _exact variant level_ or at the _codon/exon level_
	 
 - **TIER 3: Variants of uncertain clinical significance** -
   - For SNVs/InDels, this includes other coding variants found in oncogenes or tumor suppressor genes, yet _not_ linked to any known predictive, prognostic, or diagnostic biomarkers in the [CIViC database](http://civic.genome.wustl.edu) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers)
   
<br>

#### Oncogenicity classification

[@Horak2022-uh]

Included in the report is also a complete list of **all oncogenes subject to amplifications**, **tumor suppressor genes subject to homozygous deletions**, and **other drug targets subject to amplifications**

<br>

#### Mutational signatures

The set of somatic mutations observed in a tumor reflects the varied mutational processes that have been active during its life history, providing insights into the routes taken to carcinogenesis. Exogenous mutagens, such as tobacco smoke and ultraviolet light, and endogenous processes, such as APOBEC enzymatic family functional activity or DNA mismatch repair deficiency, result in characteristic patterns of mutation. There is growing evidence that mutational signatures can explain therapeutic response [@Brady2022-hr; @Levatic2022-hk].

In PCGR, we apply the **MutationalPatterns** package [@Blokzijl2018-nc] to identify the contribution of _known mutational signatures_ in a single tumor sample. Specifically, we apply [MutationalPatterns](https://github.com/UMCUGenetics/MutationalPatterns) to optimally reconstruct the observed spectrum of mutations through a [reference collection of known mutational signatures](https://www.ncbi.nlm.nih.gov/pubmed/32025018). By default, we restrict the signatures in the reference collection to those already observed in the tumor type in question (i.e. from large-scale <i>de novo</i> signature extraction on ICGC tumor samples). 

#### Tumor mutational burden (TMB)

Tumor mutational load or mutational burden is a measure of the number of mutations within a tumor genome, defined as the total number of mutations per coding area of a tumour genome. TMB may serve as a proxy for determining the number of neoantigens per tumor, which in turn may have implications for response to immunotherapy. PCGR estimates TMB according to three different schemes (one to be chosen as displayed in the report):

  1) __TMB_coding_silent__: the same approach as was outlined in a large-scale study of TMB ([Chalmers et al., 2017](https://www.ncbi.nlm.nih.gov/pubmed/28420421)), i.e. counting all somatic base substitutions and indels in the protein-coding regions of the sequencing assay, including those at canonical splice sites, and including synonymous alterations. 
  2) __TMB_coding_non_silent__: counting all somatic base substitutions and indels in the protein-coding regions of the sequencing assay, including those at canonical splice sites, but excluding synonymous alterations.
  3) __TMB_missense_only__: missense (non-synonymous) variants variants only, i.e. as employed by [Fernandez et al., 2019](https://www.ncbi.nlm.nih.gov/pubmed/31475242)

Numbers obtained with 1), 2) or 3) are next divided by the coding target size of the sequencing assay.

<br>

#### MSI classification

Microsatellite instability (MSI) is the result of impaired DNA mismatch repair and constitutes a cellular phenotype of clinical significance in many cancer types, most prominently colorectal cancers, stomach cancers, endometrial cancers, and ovarian cancers ([Cortes-Ciriano et al., 2017](https://www.ncbi.nlm.nih.gov/pubmed/28585546)). We have built a statistical MSI classifier from somatic mutation profiles that separates _MSI.H_ (MSI-high) from _MSS_ (MS stable) tumors. The MSI classifier was trained using 999 exome-sequenced TCGA tumor samples with known MSI status (i.e. assayed from mononucleotide markers), and obtained a [positive predictive value](https://en.wikipedia.org/wiki/Positive_and_negative_predictive_values#Positive_predictive_value) of 100% and a [negative predictive value](https://en.wikipedia.org/wiki/Positive_and_negative_predictive_values#Negative_predictive_value) of 99.4% on an independent test set of 427 samples. Details of the MSI classification approach can be found <a href="https://rpubs.com/sigven/msi2020_v2" target="_blank">here</a>.

Note that the MSI classifier is applied only for WGS/WES tumor-control sequencing assays.

<br>

#### Kataegis

Kataegis describes a pattern of localized hypermutations identified in some cancer genomes, in which a large number of highly-patterned basepair mutations occur in a small region of DNA (ref Wikipedia). Kataegis is prevalently seen among breast cancer patients, and it is also exists in lung cancers, cervical, head and neck, and bladder cancers, as shown in the results from tracing APOBEC mutation signatures (ref Wikipedia). PCGR implements the kataegis detection algorithm outlined in the [KataegisPortal R package](https://github.com/MeichunCai/KataegisPortal). 

Explanation of key columns in the resulting table of potential kataegis events: 

* __weight.C>X__: proportion of C>X mutations
* __confidence__: confidence degree of potential kataegis events (range: 0 to 3)
    - 0 - a hypermutation with weight.C>X < 0.8; 
    - 1 - one hypermutation with weight.C>X >= 0.8 in a chromosome; 
    - 2 - two hypermutations with weight.C>X >= 0.8 in a chromosome; 
    - 3 - high confidence with three or more hypermutations with weight.C>X >= 0.8 in a chromosome)

<br>

#### Germline findings

For PCGR reports that are fueled with CPSR report contents (JSON), we here list the main findings from the CPSR report, i.e. the collection of __Pathogenic/Likely Pathogenic/VUS variants__ (ClinVar and novel CPSR-classified variants). We also show whether any of the query variants are associated with established biomarker evidence items with respect to cancer predisposition, prognosis, therapeutic regimens etc.

<br>

#### Clinical trials

Each report is provided with a list of trials for the tumor type in question, where we limit the trials listed to ongoing or forthcoming trials with a "molecular focus" (presence of molecular biomarkers in inclusion criteria, targeted drugs as interventions etc.). Recognition of biomarkers in trials is conducted through an in-house text mining procedure. 

Note that the trials have currently not been subject to any matching with respect to the molecular profile of the tumor, trials are thus basically unprioritized, and have to be explored interactively by the user in order to discover relevant trials. 

<br>

:::
