---
title: "Installation"
output: rmarkdown::html_document

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(comment = "", collapse = TRUE)
```


```{r load_pkgs, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
require(glue, include.only = "glue")
```

```{r vars, echo=FALSE}
Sys.setenv(VEP_VERSION = "112")
Sys.setenv(PCGR_VERSION = "1.4.1.9014")
Sys.setenv(BUNDLE_VERSION = "20240612")
VEP_VERSION <- Sys.getenv("VEP_VERSION")
PCGR_VERSION <- Sys.getenv("PCGR_VERSION")
BUNDLE_VERSION <- Sys.getenv("BUNDLE_VERSION")
```

```{r funcs, echo=FALSE}
bundle_link <- function(hg) {
  v <- BUNDLE_VERSION
  glue("https://insilico.hpc.uio.no/pcgr/pcgr_ref_data.{v}.{hg}.tgz")
}
```


## Data

PCGR requires the following data:

- Sample-specific inputs (e.g. somatic variant calls in VCF format)
- Reference bundle (e.g. CIViC, CGI, TCGA)
- Ensembl VEP data cache

PCGR supports the GRCh37 and GRCh38 human genome assemblies. All the data above
need to match the chosen assembly.

### 1. Reference Bundle

Reference bundles are generated semi-automatically (by the PCGR author) and
are versioned based on their release date. Keep in mind that the bundles support
only certain Ensembl VEP versions. The genome-specific bundles
(**v`r BUNDLE_VERSION`**) can be downloaded directly from below (size: ~5G):

| Assembly | Download Link             |
|----------|---------------------------|
| GRCh38   | `r bundle_link("grch38")` |
| GRCh37   | `r bundle_link("grch37")` |

**Tip**: The `data/grch3x/.PCGR_BUNDLE_VERSION` file within the downloaded bundle
indicates the bundle version for reporting purposes.

#### Bash Example

```{bash echo=FALSE}
echo "BUNDLE_VERSION=\"${BUNDLE_VERSION}\""
```

```bash
GENOME="grch38" # or "grch37"
BUNDLE="pcgr_ref_data.${BUNDLE_VERSION}.${GENOME}.tgz"
wget https://insilico.hpc.uio.no/pcgr/${BUNDLE}
gzip -dc ${BUNDLE} | tar xvf -

mkdir ${BUNDLE_VERSION}
mv data/ ${BUNDLE_VERSION}
```

### 2. VEP Cache

[VEP][vep-web] requires a data cache which is available from the Ensembl
[FTP site][ensembl-ftp] (search there for files starting with `homo_sapiens_vep_`).
We currently support Ensembl VEP **v`r VEP_VERSION`**.

**Tip**: PCGR needs to be pointed to the _parent_ directory containing
the downloaded `homo_sapiens/xyz_GRCh3x/` cache, which is usually called `.vep` if
you've followed the VEP cache [download instructions][vep-cache].

[vep-web]: https://asia.ensembl.org/info/docs/tools/vep/index.html
[ensembl-ftp]: https://ftp.ensembl.org/pub/release-112/variation/indexed_vep_cache/
[vep-cache]: https://asia.ensembl.org/info/docs/tools/vep/script/vep_cache.html#cache

#### Bash Example

```{bash echo=FALSE}
echo "VEP_VERSION=\"${VEP_VERSION}\""
```

```bash
GENOME="GRCh38" # or "GRCh37"
CACHE="homo_sapiens_vep_${VEP_VERSION}_${GENOME}.tar.gz"

wget https://ftp.ensembl.org/pub/release-${VEP_VERSION}/variation/indexed_vep_cache/${CACHE}
gzip -dc ${CACHE} | tar xvf -
```

-----------------------------

## Software

The PCGR workflow can be installed using [Docker][docker-web],
[Singularity/Apptainer][apptainer-web] or [Conda][conda-web].

[conda-web]: https://conda.io/projects/conda/en/latest/user-guide/getting-started.html
[docker-web]: https://docs.docker.com/
[apptainer-web]: https://apptainer.org/docs/user/latest/index.html

### A. Docker

The Docker image is available on [DockerHub](https://hub.docker.com/r/sigven/pcgr/tags).
Pull the latest **v`r PCGR_VERSION`** image with:

```{r echo=FALSE}
glue("docker pull sigven/pcgr:{PCGR_VERSION}")
# might need to specify platform
# docker pull --platform=amd64 sigven/pcgr:${PCGR_VERSION}
```

#### Example Run

```bash
docker container run -it --rm \
    -v /Users/you/projects/.vep:/mnt/.vep
    -v /Users/you/projects/bundle:/mnt/bundle \
    -v /Users/you/projects/pcgr_inputs:/mnt/pcgr_inputs \
    -v /Users/you/projects/pcgr_outputs:/mnt/pcgr_outputs \
    sigven/pcgr:1.4.1.9014 \
    pcgr \
      --input_vcf "/mnt/pcgr_inputs/tumor_sample.BRCA.vcf.gz" \
      --vep_dir "/mnt/.vep" \
      --refdata_dir "/mnt/bundle" \
      --output_dir "/mnt/pcgr_outputs" \
      --genome_assembly "grch38" \
      --sample_id "SampleB" \
      --assay "WGS" \
      --vcf2maf
```

### B. Singularity/Apptainer

```{r echo=FALSE}
glue("apptainer pull oras://ghcr.io/sigven/pcgr:{PCGR_VERSION}.singularity")
```



### C. Conda

There is Conda support for both Linux and macOS machines.
The following process can take anywhere from 10 up to 40 minutes when installing
from scratch, mostly depending on the user's and server's internet connection.
Most of the time is spent on downloading the `{BSgenome.Hsapiens.UCSC.hg19}` and
`{BSgenome.Hsapiens.UCSC.hg38}` R packages (which happens at the very end of the
conda environment creation).

#### Linux

```bash
# set up variables
PCGR_VERSION="1.4.1.9014"
PCGR_REPO="https://raw.githubusercontent.com/sigven/pcgr/v${PCGR_VERSION}/conda/env/lock/"
PLATFORM="linux"
# create conda envs in local directory
mkdir pcgr_conda
conda create --prefix ./pcgr_conda/pcgr --file ${PCGR_REPO}/pcgr-${PLATFORM}-64.lock
conda create --prefix ./pcgr_conda/pcgrr --file ${PCGR_REPO}/pcgrr-${PLATFORM}-64.lock
# you need to specify the directory of the conda env when using --prefix
conda activate ./pcgr_conda/pcgr
# test that it works
pcgr --version
```

#### macOS

For macOS M1 machines, you need to include `CONDA_SUBDIR=osx-64` before the
`conda create` command - see
<https://github.com/conda-forge/miniforge/issues/165#issuecomment-860233092>:

```bash
# set up variables
PCGR_VERSION="1.4.1.9014"
PCGR_REPO="https://raw.githubusercontent.com/sigven/pcgr/v${PCGR_VERSION}/conda/env/lock/"
PLATFORM="osx"
# create conda envs in local directory
mkdir pcgr_conda
CONDA_SUBDIR=osx-64 conda create --prefix ./pcgr_conda/pcgr --file ${PCGR_REPO}/pcgr-${PLATFORM}-64.lock
CONDA_SUBDIR=osx-64 conda create --prefix ./pcgr_conda/pcgrr --file ${PCGR_REPO}/pcgrr-${PLATFORM}-64.lock
# you need to specify the directory of the conda env when using --prefix
conda activate ./pcgr_conda/pcgr
# test that it works
pcgr --version
```
