---
title: "Installation"
output: rmarkdown::html_document
---

## Quick Installation

### Conda

- If you know what [conda](https://docs.conda.io/projects/conda/en/latest/user-guide/getting-started.html)
  is, you only need to run the following commands in order to install the PCGR
  _software_ requirements into a local directory:

```bash
cd /Users/you/dir4/conda
PCGR_VERSION="1.4.1.9011"
PCGR_REPO="https://raw.githubusercontent.com/sigven/pcgr/v${PCGR_VERSION}/conda/env/lock/"
PLATFORM="linux" # or "osx"

conda create --file ${PCGR_REPO}/pcgr-${PLATFORM}-64.lock --prefix ./pcgr
conda create --file ${PCGR_REPO}/pcgrr-${PLATFORM}-64.lock --prefix ./pcgrr

# you need to specify the directory of the conda env when using --prefix
conda activate ./pcgr

# test that it works
pcgr --version
```

### Data

- For downloading the reference data bundle and VEP cache, see [STEP 1](#step1) further below.

### Docker

- If you know what [Docker](https://www.docker.com/) is, instead of using the above conda method
  you can jump straight to the [PCGR Docker setup](#dockersetup) section.

<br>
<hr>
<br>

## Detailed Installation

PCGR requires:

- a __data bundle__ that contains the reference data;
- a data __cache for VEP__ (Variant Effect Predictor);
- __sample inputs__ (e.g. somatic variants in a VCF);
- and an __output directory__ to write the results to

Here's an example scenario that will be used in the following sections
(where GRCh3x can be GRCh37 or GRCh38):

- VEP cache is downloaded in `/Users/you/dir0/.vep`;
  - make sure the `.vep` directory contains `.vep/homo_sapiens/xyz_GRCh3x/`
    for a given `<xyz>` VEP release.
- data bundle downloaded in `/Users/you/dir1/data`;
  - make sure you have a `data/grch3x/.PCGR_BUNDLE_VERSION` file
- sample inputs at `/Users/you/dir2/pcgr_inputs`;
- output goes to `/Users/you/dir3/pcgr_outputs`;
- conda environments installed in `/Users/you/dir4/conda`;

<a name="step1"></a>

### STEP 1: Download data bundle and VEP cache

**A)** Download and unpack the assembly-specific reference data bundle needed for PCGR:

- [grch37 data bundle - 20240610](https://insilico.hpc.uio.no/pcgr/pcgr_ref_data.20240610.grch37.tgz) (approx 4.8Gb)
- [grch38 data bundle - 20240610](https://insilico.hpc.uio.no/pcgr/pcgr_ref_data.20240610.grch38.tgz) (approx 4.8Gb)

- Example:

```bash
GENOME="grch38" # or "grch37"
BUNDLE_VERSION="20240610"
BUNDLE="pcgr_ref_data.${BUNDLE_VERSION}.${GENOME}.tgz"

wget https://insilico.hpc.uio.no/pcgr/${BUNDLE}
gzip -dc ${BUNDLE} | tar xvf -
```

**B)** Download and unpack the VEP cache (need for VEP - Variant Effect Predictor):

- Example:

```bash
GENOME="GRCh38" # or "GRCh37"
VEP_VERSION="112"
CACHE="homo_sapiens_vep_${VEP_VERSION}_${GENOME}.tar.gz"

wget https://ftp.ensembl.org/pub/release-${VEP_VERSION}/variation/indexed_vep_cache/${CACHE}
gzip -dc ${CACHE} | tar xvf -
```

### STEP 2: Set up Conda or Docker

Step 2 depends on if you want to use Conda or Docker:

- For Conda, continue reading the [PCGR Conda setup](#condasetup).
- For Docker, skip to the [PCGR Docker setup](#dockersetup).

<a name="condasetup"></a>

### Option 1: Conda

#### a) Miniconda and conda

Download and install the Miniconda installer from <https://docs.conda.io/en/latest/miniconda.html>:

- Make sure to download the Linux or MacOSX script according to which platform you're currently on.
- Run `bash miniconda.sh` and follow the prompts (it should be okay to accept the defaults, unless you want to choose a different
  installation location than the default `~/miniconda3`).
- Exit your current terminal session and open a new one. You should now notice something like a `(base)` string as a
  prefix in your terminal prompt. This means that you're in the `base` conda environment, and you're ready to start
  installing the conda environments for PCGR.

```bash
PLATFORM="MacOSX" # or "Linux"
MINICONDA_URL="https://repo.continuum.io/miniconda/Miniconda3-latest-${PLATFORM}-x86_64.sh"
wget ${MINICONDA_URL} -O miniconda.sh && chmod +x miniconda.sh
bash miniconda.sh
```

```text
# exit terminal and open new one - you should now see:

# as of May 2024
(base) $ conda --version
conda 24.5.0
```

#### b) Create PCGR conda environments

The `conda/env/lock` directory in the PCGR codebase contains two `.lock` files which
can be used to create the required conda environments for the Python component
(`pcgr`) and the R components (`pcgrr` (and `cpsr`)). We install the conda
dependencies for these two environments in a local `conda` directory in the
following example:

```bash
cd /Users/you/dir4/conda
PLATFORM="osx-64" # or "linux-64"

PCGR_VERSION="1.4.1.9011"
PCGR_REPO="https://raw.githubusercontent.com/sigven/pcgr/v${PCGR_VERSION}/conda/env/lock/"
PLATFORM="linux" # or "osx"

conda create --prefix ./pcgr --file ${PCGR_REPO}/pcgr-${PLATFORM}-64.lock
conda create --prefix ./pcgrr --file ${PCGR_REPO}/pcgrr-${PLATFORM}-64.lock

## Alternatively, for installing in your central conda directory, use the following:
# conda create --name pcgr --file ${PCGR_CONDA_ENV_DIR}/lock/pcgr-${PLATFORM}.lock
# conda create --name pcgrr --file ${PCGR_CONDA_ENV_DIR}/lock/pcgrr-${PLATFORM}.lock

## For MacOS M1, you need to have 'CONDA_SUBDIR=osx-64' before the conda command, i.e.:
# CONDA_SUBDIR=osx-64 conda create --prefix [...] --file [...]
## See https://github.com/conda-forge/miniforge/issues/165#issuecomment-860233092
```

The above process takes 20-30 minutes when installing from scratch. Most of the time
is spent on downloading the
{BSgenome.Hsapiens.UCSC.hg19} and {BSgenome.Hsapiens.UCSC.hg38} R packages
(and yes, for simplicity we download both packages).
In the end, confirm your conda environments have been installed correctly
(notice how the paths are different to the `base` env installation after using the
`--prefix` option above):

```text
$ (base) conda env list
# conda environments:
#
base    *  /Users/you/miniconda3
pcgr       /Users/you/dir4/conda/pcgr
pcgrr      /Users/you/dir4/conda/pcgrr
```

#### c) Activate pcgr conda environment

You need to activate the `conda/pcgr` conda environment, and test that it works
correctly with e.g. `pcgr --version`:

```text
$ cd /Users/you/dir4/conda
(base) $ conda activate ./conda/pcgr
# note how the full path to the locally installed conda environment is now displayed

(/Users/you/dir4/conda) $ which pcgr
/Users/you/dir4/conda/pcgr/bin/pcgr

(/Users/you/dir4/conda) $ pcgr --version
pcgr X.X.X

(/Users/you/dir4/conda) $ which pcgrr.R
/Users/you/dir4/conda/pcgr/bin/pcgrr.R
```

You should now be all set up to run PCGR! Continue on to [an example run](running.html#example-run).

<a name="dockersetup"></a>

### Option 2: Docker

#### a) Install Docker

For installing Docker, follow the instructions at <https://docs.docker.com/engine/install/>
for your Linux or MacOSX machine.

#### b) Download PCGR Docker Image

- Pull the [PCGR Docker image](https://hub.docker.com/r/sigven/pcgr/tags) from
  DockerHub with: `docker pull sigven/pcgr:X.X.X`

#### c) Run PCGR Docker Container

If you are familiar with working with Docker volumes (<https://docs.docker.com/storage/volumes/>)
you can run PCGR using Docker instead of conda using the `-v <host>:<container>` Docker option.
You'll need to map your PCGR inputs to Docker container paths.

For example, say you have the input VCF `sampleX.vcf.gz` stored in the
directory `/Users/you/project1`. You would need to supply Docker with a
`--volume` (or `-v`) option mapping the directory of that VCF with
a directory inside the Docker container, e.g. `/home/input_vcf_dir`.
That would become: `-v /Users/you/project1:/home/input_vcf_dir`
(note the `:` separating your directory from the container's directory).

Then your command would look something like this:

```bash
docker container run -it --rm \
    -v /Users/you/dir0/vep:/root/vep
    -v /Users/you/dir1/data:/root/pcgr_refdata \
    -v /Users/you/dir2/pcgr_inputs:/root/pcgr_inputs \
    -v /Users/you/dir3/pcgr_outputs:/root/pcgr_outputs \
    sigven/pcgr:1.4.1.9011 \
    pcgr \
      --input_vcf "/root/pcgr_inputs/tumor_sample.BRCA.vcf.gz" \
      --vep_dir "/root/vep/.vep" \
      --refdata_dir "/root/pcgr_refdata" \
      --output_dir "/root/pcgr_outputs" \
      --genome_assembly "grch38" \
      --sample_id "SampleB" \
      --assay "WGS" \
      --vcf2maf
```

- Note the path mappings. You're using the _container_ paths in the
  command, not the _host_ (your machine's) paths.
