---
title: "Variant classification"
output: rmarkdown::html_document
---

## Oncogenicity

PCGR evaluates somatic aberrations (SNV/InDels) for oncogenic potential through an implementation of standard operating procedures proposed by [VICC/CGC/ClinGen](https://pubmed.ncbi.nlm.nih.gov/35101336/). Here, various properties of the variants and genes affected are assigned criteria-specific scores, both negative and positive, pending on whether the properties support an oncogenic or benign variant nature. Criteria-specific scores are in turn aggregated towards an overall oncogenicity score per variant.

Note that all properties/criteria provided in the SOP's are _not_ readily implemented in PCGR, specifically the ones requiring manual curation or expert review (i.e. experimental oncogenic variant evidence, requiring support from _in vitro_ or _in vivo_ functional studies (criteria _OM1/OS1_)). This implies that some variants interrogated by PCGR may not be classified as oncogenic, even though they could be classified as such with more functional evidence available. Considering the nature of our current implementation, we have thus also adopted slightly different score thresholds for variant classifications to those proposed originally by [Horak et al., 2022](https://pubmed.ncbi.nlm.nih.gov/35101336/). We are working to further improve the oncogenicity classification in PCGR, and welcome feedback on this matter.

Note also that for somatic copy number aberrations, we showcase potential oncogenic events as **proto-oncogenes subject to amplifications** (where level of amplification is configurable by the user), as well as**tumor suppressor genes subject to homozygous deletions**.

The following criteria are currently used for oncogenicity classification in PCGR:

 * _CLINGEN_VICC_SBVS1_ - Very high MAF (> 0.05 in gnomAD - any five major continental populations)
 * _CLINGEN_VICC_SBS1_ - High MAF (> 0.01 in gnomAD - any five major continental populations)
 * _CLINGEN_VICC_SBP1_ - _Insilico_ support for a benign effect on the gene or gene product (multiple lines of evidence (>= 8 algorithms) from dbNSFP)
 * _CLINGEN_VICC_SBP2_ - Silent and intronic changes outside of the consensus splice site (VEP consequence)
 * _CLINGEN_VICC_OVS1_ - Null variant - in a _bona fide_ tumor suppressor gene (predicted as LoF in tumor suppressors from CGC/NCG/CancerMine)
 * _CLINGEN_VICC_OS1_ - Same amino acid change as previously established oncogenic variant (ClinVar)
 * _CLINGEN_VICC_OS3_ - Located in a mutation hotspot, >= 50 samples with a variant at amino acid position, >= 10 samples with same amino acid change (cancerhotspots.org)
 * _CLINGEN_VICC_OM1_ - Located in a presumably critical site of functional domain - here, this is implemented through indirect evidence from overlap with known predictive (drug sensitivity/resistance) loci 
 * _CLINGEN_VICC_OM2_ - Protein length changes from in-frame dels/ins in known oncogene/tumor suppressor genes or stop-loss variants in a tumor suppressor gene (tumor suppressors/oncogenes from CGC/NCG/CancerMine)
 * _CLINGEN_VICC_OM3_ - Located in a mutation hotspot, < 50 samples with a variant at amino acid position, >= 10 samples with same amino acid change (cancerhotspots.org)
 * _CLINGEN_VICC_OP1_ - _Insilico_ support for a damaging effect on the gene or gene product (multiple lines of evidence (>= 8 algorithms) from dbNSFP)
 * _CLINGEN_VICC_OP3_ - Located in a mutation hotspot, < 10 samples with the same amino acid change (cancerhotspots.org)
 * _CLINGEN_VICC_OP4_ - Absent from controls (gnomAD) / very low MAF ( < 0.0001 in all five major subpopulations)
    
## Actionability

PCGR prioritizes and evaluates variants according to clinical actionability. Currently, PCGR implements its tier classification framework along the proposed AMP/ASCO/CAP guidelines, as outlined in [Li et al., 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5707196/):

- **Tier I: Variants of strong clinical significance** - constitutes aberrations linked to predictive, prognostic, or diagnostic biomarkers in the [CIViC database](https://civicdb.org) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers) that are
	 - Found within the same tumor type/class as specified by the user, **AND**
	 - Of strong clinical evidence (i.e. approved therapies, part of guidelines, validated or discovered in late clinical trials ([CIViC evidence levels A/B](https://civic.readthedocs.io/en/latest/model/evidence/level.html)))
	 
 - **Tier II: Variants of potential clinical significance** - constitutes other aberrations linked to predictive, prognostic, or diagnostic biomarkers in the [CIViC database](https://civicdb.org) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers) that are either
	 - Of strong clinical evidence in other tumor types/classes than the one specified by the user, **OR**
	 - Of weak clinical evidence (early trials, case reports etc. ([CIViC evidence levels C/D/E](https://civic.readthedocs.io/en/latest/model/evidence/level.html)))) in the same tumor type/class as specified by the user
	 
 - **Tier III: Variants of uncertain clinical significance (SNVs/InDels only)** -
   - Other coding variants, not observed at significant allele frequencies (gnomAD MAF < 0.001), found in oncogenes or tumor suppressor genes, yet _not_ linked to any known predictive, prognostic, or diagnostic biomarkers in the [CIViC database](https://civicdb.org) and the [Cancer Biomarkers Database](https://www.cancergenomeinterpreter.org/biomarkers)

In PCGR, we skip the classification of variants into the AMP/ASCO/CAP-specified *Tier IV* (benign/likely benign variants), but rather take a more cautious approach. Specifically, for SNVs/indels that do not fall into tier I, II, or III, we classify them into *Tier V: Other coding variants*, which includes protein-coding variants in non-cancer related genes, as well as *Tier VI: Other non-coding variants*, which includes synonymous variants, intronic variants, and other variants in non-coding regions.
