

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Getting started &mdash; Personal Cancer Genome Reporter 0.9.2 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="FAQ" href="FAQ.html" />
    <link rel="prev" title="About" href="about.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Personal Cancer Genome Reporter
          

          
          </a>

          
            
            
              <div class="version">
                0.9.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#step-0-python">STEP 0: Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-1-installation-of-docker">STEP 1: Installation of Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-2-download-pcgr-and-data-bundle">STEP 2: Download PCGR and data bundle</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#latest-release">Latest release</a></li>
<li class="toctree-l3"><a class="reference internal" href="#development-version">Development version</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#step-3-input-preprocessing">STEP 3: Input preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-4-configure-your-pcgr-workflow">STEP 4: Configure your PCGR workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-5-run-example">STEP 5: Run example</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation_resources.html">PCGR annotation resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="tier_systems.html">Tier models</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">CHANGELOG</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Personal Cancer Genome Reporter</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Getting started</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/getting_started.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="getting-started">
<h1>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h1>
<div class="section" id="step-0-python">
<h2>STEP 0: Python<a class="headerlink" href="#step-0-python" title="Permalink to this headline">¶</a></h2>
<p>An installation of Python (version <em>3.6</em>) is required to run PCGR. Check
that Python is installed by typing <code class="docutils literal notranslate"><span class="pre">python</span> <span class="pre">--version</span></code> in your terminal
window.</p>
<p><strong>IMPORTANT NOTE</strong>: STEP 1 &amp; 2 outline installation guidelines for
running PCGR with Docker. If you want to install and run PCGR without
the use of Docker (i.e. through Conda), follow <a class="reference external" href="https://github.com/sigven/pcgr/blob/master/install_no_docker/README.md">these
instructions</a></p>
</div>
<div class="section" id="step-1-installation-of-docker">
<h2>STEP 1: Installation of Docker<a class="headerlink" href="#step-1-installation-of-docker" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://docs.docker.com/engine/installation/">Install the Docker
engine</a> on your
preferred platform</p>
<ul class="simple">
<li><p>installing <a class="reference external" href="https://docs.docker.com/engine/installation/linux/">Docker on
Linux</a></p></li>
<li><p>installing <a class="reference external" href="https://docs.docker.com/engine/installation/mac/">Docker on Mac
OS</a></p></li>
<li><p>NOTE: We have not yet been able to perform enough testing on the
Windows platform, and we have received feedback that particular
versions of Docker/Windows do not work with PCGR (an example being
<a class="reference external" href="https://github.com/docker/toolbox/issues/607">mounting of data
volumes</a>)</p></li>
</ul>
</li>
<li><p>Test that Docker is running, e.g. by typing <code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">ps</span></code> or
<code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">images</span></code> in the terminal window</p></li>
<li><p>Adjust the computing resources dedicated to the Docker, i.e.:</p>
<ul class="simple">
<li><p>Memory: minimum 6GB (this may need an increase for large input
files)</p></li>
<li><p>CPUs: minimum 4</p></li>
<li><p><a class="reference external" href="https://docs.docker.com/docker-for-mac/#advanced">How to - Mac OS
X</a></p></li>
</ul>
</li>
</ol>
</div>
<div class="section" id="step-2-download-pcgr-and-data-bundle">
<h2>STEP 2: Download PCGR and data bundle<a class="headerlink" href="#step-2-download-pcgr-and-data-bundle" title="Permalink to this headline">¶</a></h2>
<div class="section" id="latest-release">
<h3>Latest release<a class="headerlink" href="#latest-release" title="Permalink to this headline">¶</a></h3>
<ol class="loweralpha simple">
<li><p>Download and unpack the <a class="reference external" href="https://github.com/sigven/pcgr/releases/tag/v0.9.2">latest software release
(0.9.2)</a></p></li>
<li><p>Download and unpack the assembly-specific data bundle in the PCGR
directory</p></li>
</ol>
<ul>
<li><p><a class="reference external" href="http://insilico.hpc.uio.no/pcgr/pcgr.databundle.grch37.20210627.tgz">grch37 data bundle -
20210627</a>
(approx 20Gb)</p></li>
<li><p><a class="reference external" href="http://insilico.hpc.uio.no/pcgr/pcgr.databundle.grch38.20210627.tgz">grch38 data bundle -
20210627</a>
(approx 21Gb)</p>
<ul class="simple">
<li><p><em>Unpacking</em>:
<code class="docutils literal notranslate"><span class="pre">gzip</span> <span class="pre">-dc</span> <span class="pre">pcgr.databundle.grch37.YYYYMMDD.tgz</span> <span class="pre">|</span> <span class="pre">tar</span> <span class="pre">xvf</span> <span class="pre">-</span></code></p></li>
</ul>
<p>A <em>data/</em> folder within the <em>pcgr-X.X</em> software folder should now
have been produced</p>
</li>
</ul>
<ol class="loweralpha simple" start="3">
<li><p>Pull the <a class="reference external" href="https://hub.docker.com/r/sigven/pcgr/">PCGR Docker image
(0.9.2)</a> from DockerHub
(approx 5.0Gb):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">pull</span> <span class="pre">sigven/pcgr:0.9.2</span></code> (PCGR annotation engine)</p></li>
</ul>
</li>
</ol>
</div>
<div class="section" id="development-version">
<h3>Development version<a class="headerlink" href="#development-version" title="Permalink to this headline">¶</a></h3>
<ol class="loweralpha simple">
<li><p>Clone the PCGR GitHub repository (includes run script and default
configuration file): <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">clone</span> <span class="pre">https://github.com/sigven/pcgr.git</span></code></p></li>
<li><p>Download and unpack the latest data bundles in the PCGR directory</p></li>
</ol>
<ul class="simple">
<li><p><a class="reference external" href="http://insilico.hpc.uio.no/pcgr/pcgr.databundle.grch37.20210627.tgz">grch37 data bundle -
20210627</a>
(approx 20Gb)</p></li>
<li><p><a class="reference external" href="http://insilico.hpc.uio.no/pcgr/pcgr.databundle.grch38.20210627.tgz">grch38 data bundle -
20210627</a>
(approx 21Gb)</p></li>
<li><p><em>Unpacking</em>:
<code class="docutils literal notranslate"><span class="pre">gzip</span> <span class="pre">-dc</span> <span class="pre">pcgr.databundle.grch37.YYYYMMDD.tgz</span> <span class="pre">|</span> <span class="pre">tar</span> <span class="pre">xvf</span> <span class="pre">-</span></code></p></li>
</ul>
<ol class="loweralpha simple" start="3">
<li><p>Pull the <a class="reference external" href="https://hub.docker.com/r/sigven/pcgr/">PCGR Docker image
(dev)</a> from DockerHub
(approx 5.0Gb):</p></li>
</ol>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">pull</span> <span class="pre">sigven/pcgr:dev</span></code> (PCGR annotation engine)</p></li>
</ul>
</div>
</div>
<div class="section" id="step-3-input-preprocessing">
<h2>STEP 3: Input preprocessing<a class="headerlink" href="#step-3-input-preprocessing" title="Permalink to this headline">¶</a></h2>
<p>The PCGR workflow accepts two types of input files:</p>
<ul class="simple">
<li><p>An unannotated, single-sample VCF file (&gt;= v4.2) with called somatic
variants (SNVs/InDels)</p></li>
<li><p>A copy number segment file</p></li>
</ul>
<p>PCGR can be run with either or both of the two input files present.</p>
<ul class="simple">
<li><p>We <strong>strongly</strong> recommend that the input VCF is compressed and
indexed using <a class="reference external" href="http://www.htslib.org/doc/tabix.html">bgzip</a> and
<a class="reference external" href="http://www.htslib.org/doc/tabix.html">tabix</a></p></li>
<li><p>If the input VCF contains multi-allelic sites, these will be subject
to <a class="reference external" href="http://genome.sph.umich.edu/wiki/Vt#Decompose">decomposition</a></p></li>
<li><p>Variants used for reporting should be designated as ‘PASS’ in the VCF
FILTER column</p></li>
</ul>
<p>The tab-separated values file with copy number aberrations <strong>MUST</strong>
contain the following four columns:</p>
<ul class="simple">
<li><p>Chromosome</p></li>
<li><p>Start</p></li>
<li><p>End</p></li>
<li><p>Segment_Mean</p></li>
</ul>
<p>Here, <em>Chromosome</em>, <em>Start</em>, and <em>End</em> denote the chromosomal segment,
and <strong>Segment_Mean</strong> denotes the log(2) ratio for a particular segment,
which is a common output of somatic copy number alteration callers. Note
that coordinates must be <strong>one-based</strong> (i.e. chromosomes start at 1, not
0). Below shows the initial part of a copy number segment file that is
formatted correctly according to PCGR’s requirements:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Chromosome</span>  <span class="n">Start</span>   <span class="n">End</span> <span class="n">Segment_Mean</span>
<span class="mi">1</span> <span class="mi">3218329</span> <span class="mi">3550598</span> <span class="mf">0.0024</span>
<span class="mi">1</span> <span class="mi">3552451</span> <span class="mi">4593614</span> <span class="mf">0.1995</span>
<span class="mi">1</span> <span class="mi">4593663</span> <span class="mi">6433129</span> <span class="o">-</span><span class="mf">1.0277</span>
</pre></div>
</div>
</div>
<div class="section" id="step-4-configure-your-pcgr-workflow">
<h2>STEP 4: Configure your PCGR workflow<a class="headerlink" href="#step-4-configure-your-pcgr-workflow" title="Permalink to this headline">¶</a></h2>
<p>The PCGR software comes with a range of options to configure the report
and analysis, depending on the type of sequencing assay
(<strong>tumor-control</strong> vs. <strong>tumor_only</strong>), types of data available
(SNVs/InDels, copy number aberration) etc.</p>
<ul class="simple">
<li><p><strong>IMPORTANT</strong>: Proper designation of VCF INFO tags that denote
variant depth/allelic fraction is critical to make the report as
comprehensive as possible (options <code class="docutils literal notranslate"><span class="pre">--tumor_dp_tag</span></code>,
<code class="docutils literal notranslate"><span class="pre">--tumor_af_tag</span></code>,<code class="docutils literal notranslate"><span class="pre">--control_dp_tag</span></code>,<code class="docutils literal notranslate"><span class="pre">--control_af_tag</span></code>, and
<code class="docutils literal notranslate"><span class="pre">--call_conf_tag</span></code>)</p>
<ul>
<li><p>If these items are not available/properly set, the report contents
will be less informative <strong>AND</strong> it will not be possible to preset
thresholds for variant depth/allelic fraction</p></li>
</ul>
</li>
</ul>
<p>Furthermore, to make the report as complete as possible, take a note of
the following arguments which can be used in the report generation:</p>
<ul class="simple">
<li><p>Tumor purity estimate (<code class="docutils literal notranslate"><span class="pre">--tumor_ploidy</span></code>, for display only)</p></li>
<li><p>Tumor ploidy estimate (<code class="docutils literal notranslate"><span class="pre">--tumor_purity</span></code>, for display only)</p></li>
<li><p>Type of sequencing assay (<code class="docutils literal notranslate"><span class="pre">--assay</span></code>)</p></li>
<li><p>Cell line sample (<code class="docutils literal notranslate"><span class="pre">--cell_line</span></code>, for display only)</p></li>
<li><p>Coding target size of sequencing assay (<code class="docutils literal notranslate"><span class="pre">--target_size_mb</span></code>)</p></li>
</ul>
</div>
<div class="section" id="step-5-run-example">
<h2>STEP 5: Run example<a class="headerlink" href="#step-5-run-example" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>usage:
pcgr.py -h [options]
--input_vcf &lt;INPUT_VCF&gt;
--pcgr_dir &lt;PCGR_DIR&gt;
--output_dir &lt;OUTPUT_DIR&gt;
--genome_assembly &lt;GENOME_ASSEMBLY&gt;
--sample_id &lt;SAMPLE_ID&gt;


Personal Cancer Genome Reporter (PCGR) workflow for clinical interpretation of somatic nucleotide variants and copy number aberration segments

Required arguments:
--input_vcf INPUT_VCF
                VCF input file with somatic variants in tumor sample, SNVs/InDels
--pcgr_dir PCGR_DIR   PCGR base directory with accompanying data directory, e.g. ~/pcgr-0.9.2
--output_dir OUTPUT_DIR
                Output directory
--genome_assembly {grch37,grch38}
                Human genome assembly build: grch37 or grch38
--sample_id SAMPLE_ID
                Tumor sample/cancer genome identifier - prefix for output files

vcfanno options:
--vcfanno_n_proc VCFANNO_N_PROC
                Number of vcfanno processes (option &#39;-p&#39; in vcfanno), default: 4

VEP options:
--vep_n_forks VEP_N_FORKS
                Number of forks (option &#39;--fork&#39; in VEP), default: 4
--vep_buffer_size VEP_BUFFER_SIZE
                Variant buffer size (variants read into memory simultaneously, option &#39;--buffer_size&#39; in VEP)
                - set lower to reduce memory usage, default: 100
--vep_pick_order VEP_PICK_ORDER
                Comma-separated string of ordered transcript/variant properties for selection of primary variant consequence
                ( option &#39;--pick_order&#39; in VEP), default: canonical,appris,biotype,ccds,rank,tsl,length,mane
--vep_no_intergenic   Skip intergenic variants during processing (option &#39;--no_intergenic&#39; in VEP), default: False
--vep_regulatory      Add VEP regulatory annotations (option &#39;--regulatory&#39; )or non-coding interpretation, default: False

Tumor mutational burden (TMB) and MSI options:
--target_size_mb TARGET_SIZE_MB
                For mutational burden analysis - approximate protein-coding target size in Mb of sequencing assay (default: 34 (WES/WGS))
--estimate_tmb        Estimate tumor mutational burden from the total number of somatic mutations and target region size, default: False
--estimate_msi_status
                Predict microsatellite instability status from patterns of somatic mutations/indels, default: False
--tmb_algorithm {all_coding,nonsyn}
                Method for calculation of TMB, all coding variants (Chalmers et al., Genome Medicine, 2017), or non-synonymous variants only, default: all_coding

Mutatonal signature options:
--estimate_signatures
                Estimate relative contributions of reference mutational signatures in query sample and detect potential kataegis events, default: False
--min_mutations_signatures MIN_MUTATIONS_SIGNATURES
                Minimum number of SNVs required for reconstruction of mutational signatures (SBS) by MutationalPatterns (default: 200, minimum n = 100)
--all_reference_signatures
                Use all reference mutational signatures (SBS, n = 67) in signature reconstruction rather than only those already attributed to the tumor type (default: False)
--include_artefact_signatures
                Include sequencing artefacts in the collection of reference signatures (default: False

Tumor-only options:
--tumor_only          Input VCF comes from tumor-only sequencing, calls will be filtered for variants of germline origin, (default: False)
--cell_line           Input VCF comes from tumor cell line sequencing (requires --tumor_only), calls will be filtered for variants of germline origin, (default: False)
--pon_vcf PON_VCF     VCF file with germline calls from Panel of Normals (PON) - i.e. blacklisted variants, (default: None)
--maf_onekg_eur MAF_ONEKG_EUR
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - European pop, default: 0.002)
--maf_onekg_amr MAF_ONEKG_AMR
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - Ad Mixed American pop, default: 0.002)
--maf_onekg_afr MAF_ONEKG_AFR
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - African pop, default: 0.002)
--maf_onekg_eas MAF_ONEKG_EAS
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - East Asian pop, default: 0.002)
--maf_onekg_sas MAF_ONEKG_SAS
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - South Asian pop, default: 0.002)
--maf_onekg_global MAF_ONEKG_GLOBAL
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold (1000 Genomes Project - global pop, default: 0.002)
--maf_gnomad_nfe MAF_GNOMAD_NFE
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - European (non-Finnish), default: 0.002)
--maf_gnomad_asj MAF_GNOMAD_ASJ
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - Ashkenazi Jewish, default: 0.002)
--maf_gnomad_fin MAF_GNOMAD_FIN
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - European (Finnish), default: 0.002)
--maf_gnomad_oth MAF_GNOMAD_OTH
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - Other, default: 0.002)
--maf_gnomad_amr MAF_GNOMAD_AMR
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - Latino/Admixed American, default: 0.002)
--maf_gnomad_afr MAF_GNOMAD_AFR
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - African/African-American, default: 0.002)
--maf_gnomad_eas MAF_GNOMAD_EAS
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - East Asian, default: 0.002)
--maf_gnomad_sas MAF_GNOMAD_SAS
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - South Asian, default: 0.002)
--maf_gnomad_global MAF_GNOMAD_GLOBAL
                Exclude variants in tumor (SNVs/InDels, tumor-only mode) with MAF above the given percent threshold, (gnomAD - global population, default: 0.002)
--exclude_pon         Exclude variants occurring in PoN (Panel of Normals, if provided as VCF (--pon_vcf), default: False)
--exclude_likely_hom_germline
                Exclude likely homozygous germline variants (100 pct allelic fraction for alternate allele in tumor, very unlikely somatic event, default: False)
--exclude_likely_het_germline
                Exclude likely heterozygous germline variants (40-60 pct allelic fraction, AND presence in dbSNP + gnomAD, AND not existing as somatic event in COSMIC/TCGA, default: False)
--exclude_dbsnp_nonsomatic
                Exclude variants found in dbSNP (only those that are NOT found in ClinVar(somatic origin)/DoCM/TCGA/COSMIC, defult: False)
--exclude_nonexonic   Exclude non-exonic variants, defult: False)

Allelic support options:
--tumor_dp_tag TUMOR_DP_TAG
                Specify VCF INFO tag for sequencing depth (tumor, must be Type=Integer, default: _NA_
--tumor_af_tag TUMOR_AF_TAG
                Specify VCF INFO tag for variant allelic fraction (tumor,  must be Type=Float, default: _NA_
--control_dp_tag CONTROL_DP_TAG
                Specify VCF INFO tag for sequencing depth (control, must be Type=Integer, default: _NA_
--control_af_tag CONTROL_AF_TAG
                Specify VCF INFO tag for variant allelic fraction (control, must be Type=Float, default: _NA_
--call_conf_tag CALL_CONF_TAG
                Specify VCF INFO tag for somatic variant call confidence (must be categorical, e.g. Type=String, default: _NA_
--tumor_dp_min TUMOR_DP_MIN
                If VCF INFO tag for sequencing depth (tumor) is specified and found, set minimum required depth for inclusion in report (default: 0)
--tumor_af_min TUMOR_AF_MIN
                If VCF INFO tag for variant allelic fraction (tumor) is specified and found, set minimum required AF for inclusion in report (default: 0)
--control_dp_min CONTROL_DP_MIN
                If VCF INFO tag for sequencing depth (control) is specified and found, set minimum required depth for inclusion in report (default: 0)
--control_af_max CONTROL_AF_MAX
                If VCF INFO tag for variant allelic fraction (control) is specified and found, set maximum tolerated AF for inclusion in report (default: 1)

Other options:
--input_cna INPUT_CNA
                Somatic copy number alteration segments (tab-separated values)
--logr_gain LOGR_GAIN
                Log ratio-threshold for regions containing copy number gains/amplifications (default: 0.8)
--logr_homdel LOGR_HOMDEL
                Log ratio-threshold for regions containing homozygous deletions (default: -0.8)
--cna_overlap_pct CNA_OVERLAP_PCT
                Mean percent overlap between copy number segment and gene transcripts for reporting of gains/losses in tumor suppressor genes/oncogenes, (default: 50)
--tumor_site TSITE    Optional integer code to specify primary tumor type/site of query sample,
                choose any of the following identifiers:
                1 = Adrenal Gland
                2 = Ampulla of Vater
                3 = Biliary Tract
                4 = Bladder/Urinary Tract
                5 = Bone
                6 = Breast
                7 = Cervix
                8 = CNS/Brain
                9 = Colon/Rectum
                10 = Esophagus/Stomach
                11 = Eye
                12 = Head and Neck
                13 = Kidney
                14 = Liver
                15 = Lung
                16 = Lymphoid
                17 = Myeloid
                18 = Ovary/Fallopian Tube
                19 = Pancreas
                20 = Peripheral Nervous System
                21 = Peritoneum
                22 = Pleura
                23 = Prostate
                24 = Skin
                25 = Soft Tissue
                26 = Testis
                27 = Thymus
                28 = Thyroid
                29 = Uterus
                30 = Vulva/Vagina
                (default: 0 - any tumor type)
--tumor_purity TUMOR_PURITY
                Estimated tumor purity (between 0 and 1, (default: None)
--tumor_ploidy TUMOR_PLOIDY
                Estimated tumor ploidy (default: None)
--cpsr_report CPSR_REPORT
                CPSR report file (Gzipped JSON - file ending with &#39;cpsr.&lt;genome_assembly&gt;.json.gz&#39; -  germline report of patient&#39;s blood/control sample
--vcf2maf             Generate a MAF file for input VCF using https://github.com/mskcc/vcf2maf (default: False)
--show_noncoding      List non-coding (i.e. non protein-altering) variants in report, default: False
--assay {WES,WGS,TARGETED}
                Type of DNA sequencing assay performed for input data (VCF) default: WES
--include_trials      (Beta) Include relevant ongoing or future clinical trials, focusing on studies with molecularly targeted interventions
--preserved_info_tags PRESERVED_INFO_TAGS
                Comma-separated string of VCF INFO tags from query VCF that should be kept in PCGR output TSV file
--report_theme {default,cerulean,journal,flatly,readable,spacelab,united,cosmo,lumen,paper,sandstone,simplex,yeti}
                Visual report theme (rmarkdown)
--report_nonfloating_toc
                Do not float the table of contents (TOC) in output report (rmarkdown), default: False
--force_overwrite     By default, the script will fail with an error if any output file already exists. You can force the overwrite of existing result files by using this flag, default: False
--version             show program&#39;s version number and exit
--basic               Run functional variant annotation on VCF through VEP/vcfanno, omit other analyses (i.e. Tier assignment/MSI/TMB/Signatures etc. and report generation (STEP 4), default: False
--no_vcf_validate     Skip validation of input VCF with Ensembl&#39;s vcf-validator, default: False
--docker_uid DOCKER_USER_ID
                Docker user ID. default is the host system user ID. If you are experiencing permission errors, try setting this up to root (`--docker-uid root`)
--no_docker           Run the PCGR workflow in a non-Docker mode (see install_no_docker/ folder for instructions)
--debug               Print full Docker commands to log, default: False
</pre></div>
</div>
<p>The <em>examples</em> folder contain input VCF files from two tumor samples
sequenced within TCGA (<strong>GRCh37</strong> only). It also contains a PCGR
configuration file customized for these VCFs. A report for a colorectal
tumor case can be generated by running the following command in your
terminal window:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">~/</span><span class="n">pcgr</span><span class="o">-</span><span class="mf">0.9</span><span class="o">.</span><span class="mi">2</span><span class="o">/</span><span class="n">pcgr</span><span class="o">.</span><span class="n">py</span>
<span class="o">--</span><span class="n">pcgr_dir</span> <span class="o">~/</span><span class="n">pcgr</span><span class="o">-</span><span class="mf">0.9</span><span class="o">.</span><span class="mi">2</span>
<span class="o">--</span><span class="n">output_dir</span> <span class="o">~/</span><span class="n">pcgr</span><span class="o">-</span><span class="mf">0.9</span><span class="o">.</span><span class="mi">2</span>
<span class="o">--</span><span class="n">sample_id</span> <span class="n">tumor_sample</span><span class="o">.</span><span class="n">COAD</span>
<span class="o">--</span><span class="n">tumor_dp_tag</span> <span class="n">TDP</span>
<span class="o">--</span><span class="n">tumor_af_tag</span> <span class="n">TVAF</span>
<span class="o">--</span><span class="n">call_conf_tag</span> <span class="n">TAL</span>
<span class="o">--</span><span class="n">genome_assembly</span> <span class="n">grch37</span>
<span class="o">--</span><span class="n">input_vcf</span> <span class="o">~/</span><span class="n">pcgr</span><span class="o">-</span><span class="mf">0.9</span><span class="o">.</span><span class="mi">2</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">tumor_sample</span><span class="o">.</span><span class="n">COAD</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
<span class="o">--</span><span class="n">tumor_site</span> <span class="mi">9</span>
<span class="o">--</span><span class="n">input_cna</span> <span class="o">~/</span><span class="n">pcgr</span><span class="o">-</span><span class="mf">0.9</span><span class="o">.</span><span class="mi">2</span><span class="o">/</span><span class="n">examples</span><span class="o">/</span><span class="n">tumor_sample</span><span class="o">.</span><span class="n">COAD</span><span class="o">.</span><span class="n">cna</span><span class="o">.</span><span class="n">tsv</span>
<span class="o">--</span><span class="n">tumor_purity</span> <span class="mf">0.9</span>
<span class="o">--</span><span class="n">tumor_ploidy</span> <span class="mf">2.0</span>
<span class="o">--</span><span class="n">include_trials</span>
<span class="o">--</span><span class="n">assay</span> <span class="n">WES</span>
<span class="o">--</span><span class="n">estimate_signatures</span>
<span class="o">--</span><span class="n">estimate_msi_status</span>
<span class="o">--</span><span class="n">estimate_tmb</span>
<span class="o">--</span><span class="n">no_vcf_validate</span>
</pre></div>
</div>
<p>This command will run the Docker-based PCGR workflow and produce the
following output files in the <em>examples</em> folder:</p>
<ol class="arabic simple">
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.html</strong> - An interactive HTML
report for clinical interpretation (rmarkdown)</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.flexdb.html</strong> - An interactive
HTML report for clinical interpretation (flexdashboard)</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.vcf.gz (.tbi)</strong> - Bgzipped VCF
file with rich set of annotations for precision oncology</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.pass.vcf.gz (.tbi)</strong> - Bgzipped
VCF file with rich set of annotations for precision oncology (PASS
variants only)</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.pass.tsv.gz</strong> - Compressed
vcf2tsv-converted file with rich set of annotations for precision
oncology</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.snvs_indels.tiers.tsv</strong> -
Tab-separated values file with variants organized according to tiers
of functional relevance</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.mutational_signatures.tsv</strong> -
Tab-separated values file with information on contribution of
mutational signatures</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.json.gz</strong> - Compressed JSON dump
of HTML report content</p></li>
<li><p><strong>tumor_sample.COAD.pcgr_acmg.grch37.cna_segments.tsv.gz</strong> -
Compressed tab-separated values file with annotations of gene
transcripts that overlap with somatic copy number aberrations</p></li>
</ol>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="FAQ.html" class="btn btn-neutral float-right" title="FAQ" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="about.html" class="btn btn-neutral float-left" title="About" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-2021, Sigve Nakken

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>